---
output:
  word_document:
    path: C:/Users/Nature/Desktop/STUDY/Programming/Works/Programming-In-Learning-Bioinformatics-ZeFengZhu/Omics/Genomics/åŸºå› ç»„ä¿¡æ¯å­¦_å®éªŒè®°å½•.docx
export_on_save:
pandoc: true
---

<br /><br /><br /><br /><br /><br /><br />
<center><h1>åŸºå› ç»„ä¿¡æ¯å­¦ï¼šå®éªŒè®°å½•</h1></center>
<br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br /><br />
<p><center>å¹´çº§ï¼š2017çº§</center></p>
<p><center>ä¸“ä¸šï¼šç”Ÿç‰©ä¿¡æ¯å­¦</center></p>
<p><center>å­¦å·ï¼š1730416009</center></p>
<p><center>å§“åï¼šæœ±æ³½å³°</center></p>
<br /><br /><br /><br /><br /><br /><br /><br />
<center><h2>ç›®å½•</h2></center>

- [å®éªŒé¡¹ç›®1ï¼šåŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae1%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%a8%a1%e6%8b%9f)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84)
  - [å®éªŒæµç¨‹](#%e5%ae%9e%e9%aa%8c%e6%b5%81%e7%a8%8b)
  - [1. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿå·¥å…·ç›¸å…³æ–‡çŒ®èµ„æ–™çš„è°ƒç ”](#1-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%a8%a1%e6%8b%9f%e5%b7%a5%e5%85%b7%e7%9b%b8%e5%85%b3%e6%96%87%e7%8c%ae%e8%b5%84%e6%96%99%e7%9a%84%e8%b0%83%e7%a0%94)
    - [From pubmed](#from-pubmed)
  - [2. æŸ¥é˜… art_illumina è½¯ä»¶ä¸­çš„å¸¸ç”¨æµ‹åºå¹³å°ä¿¡æ¯](#2-%e6%9f%a5%e9%98%85-artillumina-%e8%bd%af%e4%bb%b6%e4%b8%ad%e7%9a%84%e5%b8%b8%e7%94%a8%e6%b5%8b%e5%ba%8f%e5%b9%b3%e5%8f%b0%e4%bf%a1%e6%81%af)
  - [3. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿè½¯ä»¶çš„å®‰è£…å’Œé¢„æµ‹](#3-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%a8%a1%e6%8b%9f%e8%bd%af%e4%bb%b6%e7%9a%84%e5%ae%89%e8%a3%85%e5%92%8c%e9%a2%84%e6%b5%8b)
    - [3.1 Win10å¹³å°](#31-win10%e5%b9%b3%e5%8f%b0)
    - [3.2 Ubuntu](#32-ubuntu)
    - [åŸç†](#%e5%8e%9f%e7%90%86)
    - [IO](#io)
      - [FASTQ](#fastq)
      - [SAM](#sam)
      - [STAT](#stat)
  - [4. åŸºå› ç»„æ•°æ®ä¸‹è½½](#4-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%95%b0%e6%8d%ae%e4%b8%8b%e8%bd%bd)
    - [4.1 æœç´¢æ•°æ®](#41-%e6%90%9c%e7%b4%a2%e6%95%b0%e6%8d%ae)
    - [4.2 åŸºå› ç»„ç»„è£…åºåˆ—æ•°æ®ä¸‹è½½](#42-%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%bb%84%e8%a3%85%e5%ba%8f%e5%88%97%e6%95%b0%e6%8d%ae%e4%b8%8b%e8%bd%bd)
    - [4.3 åŸºå› ç»„æµ‹åºæ•°æ®ä¸‹è½½[æ‹“å±•å†…å®¹]](#43-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae%e4%b8%8b%e8%bd%bd%e6%8b%93%e5%b1%95%e5%86%85%e5%ae%b9)
      - [4.3.1 ä¿¡æ¯](#431-%e4%bf%a1%e6%81%af)
      - [4.3.2 åˆ©ç”¨NCBI SRA Toolkitä¸‹è½½å’Œæå–è¯¥æµ‹åºæ•°æ®](#432-%e5%88%a9%e7%94%a8ncbi-sra-toolkit%e4%b8%8b%e8%bd%bd%e5%92%8c%e6%8f%90%e5%8f%96%e8%af%a5%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae)
      - [4.3.3 åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿçš„æ•°æ®æ¨¡å‹(profile)åˆ›å»º](#433-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%a8%a1%e6%8b%9f%e7%9a%84%e6%95%b0%e6%8d%ae%e6%a8%a1%e5%9e%8bprofile%e5%88%9b%e5%bb%ba)
  - [5. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ](#5-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e6%a8%a1%e6%8b%9f)
    - [5.1 ä½¿ç”¨artç³»åˆ—è½¯ä»¶, å¯¹ä¸‹è½½åŸºå› ç»„åºåˆ—è¿›è¡Œå…¨åŸºå› ç»„æµ‹åºè¿›è¡ŒåŒæœ«ç«¯æ¨¡æ‹Ÿ](#51-%e4%bd%bf%e7%94%a8art%e7%b3%bb%e5%88%97%e8%bd%af%e4%bb%b6-%e5%af%b9%e4%b8%8b%e8%bd%bd%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%ba%8f%e5%88%97%e8%bf%9b%e8%a1%8c%e5%85%a8%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e8%bf%9b%e8%a1%8c%e5%8f%8c%e6%9c%ab%e7%ab%af%e6%a8%a1%e6%8b%9f)
    - [5.2 ä½¿ç”¨åˆé€‚çš„è½¯ä»¶æˆ–å·¥å…·æ¥æŸ¥çœ‹æ¨¡æ‹Ÿè¿ç®—çš„è¾“å‡ºç»“æœæ–‡ä»¶åŠå…¶å†…å®¹ã€‚](#52-%e4%bd%bf%e7%94%a8%e5%90%88%e9%80%82%e7%9a%84%e8%bd%af%e4%bb%b6%e6%88%96%e5%b7%a5%e5%85%b7%e6%9d%a5%e6%9f%a5%e7%9c%8b%e6%a8%a1%e6%8b%9f%e8%bf%90%e7%ae%97%e7%9a%84%e8%be%93%e5%87%ba%e7%bb%93%e6%9e%9c%e6%96%87%e4%bb%b6%e5%8f%8a%e5%85%b6%e5%86%85%e5%ae%b9)
    - [5.3 ç»Ÿè®¡ä¸åŒå‚æ•°è®¾ç½®ä¸‹çš„æ¨¡æ‹Ÿç»“æœ](#53-%e7%bb%9f%e8%ae%a1%e4%b8%8d%e5%90%8c%e5%8f%82%e6%95%b0%e8%ae%be%e7%bd%ae%e4%b8%8b%e7%9a%84%e6%a8%a1%e6%8b%9f%e7%bb%93%e6%9e%9c)
      - [5.3.1 SRR800817æµ‹åºè¦†ç›–åº¦ç†è®ºå€¼](#531-srr800817%e6%b5%8b%e5%ba%8f%e8%a6%86%e7%9b%96%e5%ba%a6%e7%90%86%e8%ae%ba%e5%80%bc)
      - [5.3.2 ä¸åŒå‚æ•°è®¾ç½®ä¸‹è¦†ç›–åº¦æ•°æ®](#532-%e4%b8%8d%e5%90%8c%e5%8f%82%e6%95%b0%e8%ae%be%e7%bd%ae%e4%b8%8b%e8%a6%86%e7%9b%96%e5%ba%a6%e6%95%b0%e6%8d%ae)
- [å®éªŒé¡¹ç›®2ï¼šåºåˆ—ç»„è£…](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae2%e5%ba%8f%e5%88%97%e7%bb%84%e8%a3%85)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84-1)
  - [å®éªŒæµç¨‹](#%e5%ae%9e%e9%aa%8c%e6%b5%81%e7%a8%8b-1)
  - [1. æ•°æ®å‡†å¤‡](#1-%e6%95%b0%e6%8d%ae%e5%87%86%e5%a4%87)
    - [1.1 å‡†å¤‡åŸå§‹åŸºå› ç»„ç»„è£…åºåˆ—](#11-%e5%87%86%e5%a4%87%e5%8e%9f%e5%a7%8b%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%bb%84%e8%a3%85%e5%ba%8f%e5%88%97)
    - [1.2 åˆ©ç”¨art_illuminaæ¨¡æ‹ŸåŒæœ«ç«¯æµ‹åº](#12-%e5%88%a9%e7%94%a8artillumina%e6%a8%a1%e6%8b%9f%e5%8f%8c%e6%9c%ab%e7%ab%af%e6%b5%8b%e5%ba%8f)
    - [1.3 åˆ›å»ºå‚è€ƒåŸºå› ç»„](#13-%e5%88%9b%e5%bb%ba%e5%8f%82%e8%80%83%e5%9f%ba%e5%9b%a0%e7%bb%84)
  - [2. è´¨æ§åˆ†æ](#2-%e8%b4%a8%e6%8e%a7%e5%88%86%e6%9e%90)
  - [3. æŸ¥é˜…ä¸¤å¥—æ•°æ®çš„åˆ†æç»“æœå¹¶åˆ†æç»“æœ](#3-%e6%9f%a5%e9%98%85%e4%b8%a4%e5%a5%97%e6%95%b0%e6%8d%ae%e7%9a%84%e5%88%86%e6%9e%90%e7%bb%93%e6%9e%9c%e5%b9%b6%e5%88%86%e6%9e%90%e7%bb%93%e6%9e%9c)
    - [3.1 pair1](#31-pair1)
    - [3.2 pair2](#32-pair2)
    - [3.3 mate-pair1](#33-mate-pair1)
    - [3.4 mate-pair2](#34-mate-pair2)
  - [4. æµ‹åºæ•°æ®ä¸å‚è€ƒåŸºå› ç»„çš„æ¯”å¯¹](#4-%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae%e4%b8%8e%e5%8f%82%e8%80%83%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%9a%84%e6%af%94%e5%af%b9)
    - [4.1 åˆ©ç”¨bowtie2ï¼ŒæŠŠä¸Šè¿°ä¸¤å¥—é«˜é€šé‡æµ‹åºæ•°æ®ä¸ä¹‹å‰å‡†å¤‡å¥½çš„åŸºå› ç»„ç´¢å¼•æ–‡ä»¶è¿›è¡Œæ¯”å¯¹ï¼Œä¿ç•™æ¯”å¯¹ç»“æœ](#41-%e5%88%a9%e7%94%a8bowtie2%e6%8a%8a%e4%b8%8a%e8%bf%b0%e4%b8%a4%e5%a5%97%e9%ab%98%e9%80%9a%e9%87%8f%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae%e4%b8%8e%e4%b9%8b%e5%89%8d%e5%87%86%e5%a4%87%e5%a5%bd%e7%9a%84%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%b4%a2%e5%bc%95%e6%96%87%e4%bb%b6%e8%bf%9b%e8%a1%8c%e6%af%94%e5%af%b9%e4%bf%9d%e7%95%99%e6%af%94%e5%af%b9%e7%bb%93%e6%9e%9c)
    - [4.2 åˆ©ç”¨samtoolså¯¹æ¯”å¯¹ç»“æœå¹¶è¿›è¡Œç®€å•çš„ç»Ÿè®¡åˆ†æ](#42-%e5%88%a9%e7%94%a8samtools%e5%af%b9%e6%af%94%e5%af%b9%e7%bb%93%e6%9e%9c%e5%b9%b6%e8%bf%9b%e8%a1%8c%e7%ae%80%e5%8d%95%e7%9a%84%e7%bb%9f%e8%ae%a1%e5%88%86%e6%9e%90)
    - [4.3 è§£è¯»samtools statsçš„ç»Ÿè®¡ç»“æœ,åˆ©ç”¨plot-bamstatså·¥å…·å¯¹è¾“å‡ºçš„ç»“æœæ–‡æ¡£è¿›è¡Œå¯è§†åŒ–](#43-%e8%a7%a3%e8%af%bbsamtools-stats%e7%9a%84%e7%bb%9f%e8%ae%a1%e7%bb%93%e6%9e%9c%e5%88%a9%e7%94%a8plot-bamstats%e5%b7%a5%e5%85%b7%e5%af%b9%e8%be%93%e5%87%ba%e7%9a%84%e7%bb%93%e6%9e%9c%e6%96%87%e6%a1%a3%e8%bf%9b%e8%a1%8c%e5%8f%af%e8%a7%86%e5%8c%96)
  - [5. åºåˆ—ç»„è£…åŠç»“æœåˆ†æ](#5-%e5%ba%8f%e5%88%97%e7%bb%84%e8%a3%85%e5%8f%8a%e7%bb%93%e6%9e%9c%e5%88%86%e6%9e%90)
    - [5.1 å¯¹ä¸Šè¿°é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œç»„è£…](#51-%e5%af%b9%e4%b8%8a%e8%bf%b0%e9%ab%98%e9%80%9a%e9%87%8f%e6%b5%8b%e5%ba%8f%e6%95%b0%e6%8d%ae%e8%bf%9b%e8%a1%8c%e7%bb%84%e8%a3%85)
    - [5.2 åˆ©ç”¨Quastå°†ç»„è£…ç»“æœä¸­åŒ…å«contigså’Œscaffoldsåºåˆ—çš„æ–‡æ¡£ä¸å‚è€ƒåŸºå› ç»„è¿›è¡Œå¯¹æ¯”](#52-%e5%88%a9%e7%94%a8quast%e5%b0%86%e7%bb%84%e8%a3%85%e7%bb%93%e6%9e%9c%e4%b8%ad%e5%8c%85%e5%90%abcontigs%e5%92%8cscaffolds%e5%ba%8f%e5%88%97%e7%9a%84%e6%96%87%e6%a1%a3%e4%b8%8e%e5%8f%82%e8%80%83%e5%9f%ba%e5%9b%a0%e7%bb%84%e8%bf%9b%e8%a1%8c%e5%af%b9%e6%af%94)
    - [5.3 æŸ¥çœ‹å¹¶åˆ†ææ¯”å¯¹ç»“æœï¼Œå…³æ³¨å®é™…è¦†ç›–ç‡ç­‰é‡è¦è¯„ä¼°æŒ‡æ ‡](#53-%e6%9f%a5%e7%9c%8b%e5%b9%b6%e5%88%86%e6%9e%90%e6%af%94%e5%af%b9%e7%bb%93%e6%9e%9c%e5%85%b3%e6%b3%a8%e5%ae%9e%e9%99%85%e8%a6%86%e7%9b%96%e7%8e%87%e7%ad%89%e9%87%8d%e8%a6%81%e8%af%84%e4%bc%b0%e6%8c%87%e6%a0%87)
- [å®éªŒé¡¹ç›®3: åŸºå› ç»„æ³¨é‡Šä¹‹åŒæºæœç´¢](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae3-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b3%a8%e9%87%8a%e4%b9%8b%e5%90%8c%e6%ba%90%e6%90%9c%e7%b4%a2)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84-2)
  - [å®éªŒæµç¨‹](#%e5%ae%9e%e9%aa%8c%e6%b5%81%e7%a8%8b-2)
  - [1. æ•°æ®å‡†å¤‡åŠé¢„å¤„ç†](#1-%e6%95%b0%e6%8d%ae%e5%87%86%e5%a4%87%e5%8f%8a%e9%a2%84%e5%a4%84%e7%90%86)
    - [1.1 åŸºå› ç»„åºåˆ—](#11-%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%ba%8f%e5%88%97)
    - [1.2 å·²çŸ¥è›‹ç™½åºåˆ—](#12-%e5%b7%b2%e7%9f%a5%e8%9b%8b%e7%99%bd%e5%ba%8f%e5%88%97)
  - [2 åˆ›å»ºæœ¬åœ° BLAST æ•°æ®åº“](#2-%e5%88%9b%e5%bb%ba%e6%9c%ac%e5%9c%b0-blast-%e6%95%b0%e6%8d%ae%e5%ba%93)
  - [3å…¨åŸºå› ç»„çš„åŒæºåŸºå› æœç´¢](#3%e5%85%a8%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%9a%84%e5%90%8c%e6%ba%90%e5%9f%ba%e5%9b%a0%e6%90%9c%e7%b4%a2)
    - [3.1 ä½¿ç”¨ tblastn ç¨‹åº, æŠŠå·²çŸ¥è›‹ç™½è´¨åºåˆ—å’Œä¸Šè¿°å»ºç«‹çš„æœ¬åœ° BLAST æ•°æ®åº“è¿›è¡Œæ¯”å¯¹](#31-%e4%bd%bf%e7%94%a8-tblastn-%e7%a8%8b%e5%ba%8f-%e6%8a%8a%e5%b7%b2%e7%9f%a5%e8%9b%8b%e7%99%bd%e8%b4%a8%e5%ba%8f%e5%88%97%e5%92%8c%e4%b8%8a%e8%bf%b0%e5%bb%ba%e7%ab%8b%e7%9a%84%e6%9c%ac%e5%9c%b0-blast-%e6%95%b0%e6%8d%ae%e5%ba%93%e8%bf%9b%e8%a1%8c%e6%af%94%e5%af%b9)
    - [3.2 ä½¿ç”¨ blast92gff3.pl å’Œ blast2gff.py ç¨‹åºï¼Œåˆ†åˆ«æŠŠç»“æœè½¬æˆ GFF3 æ ¼å¼](#32-%e4%bd%bf%e7%94%a8-blast92gff3pl-%e5%92%8c-blast2gffpy-%e7%a8%8b%e5%ba%8f%e5%88%86%e5%88%ab%e6%8a%8a%e7%bb%93%e6%9e%9c%e8%bd%ac%e6%88%90-gff3-%e6%a0%bc%e5%bc%8f)
    - [3.3 æ¯”è¾ƒä¸¤ä¸ªç¨‹åºè½¬æ¢ç»“æœçš„å¼‚åŒä¹‹å¤„](#33-%e6%af%94%e8%be%83%e4%b8%a4%e4%b8%aa%e7%a8%8b%e5%ba%8f%e8%bd%ac%e6%8d%a2%e7%bb%93%e6%9e%9c%e7%9a%84%e5%bc%82%e5%90%8c%e4%b9%8b%e5%a4%84)
    - [3.4 æ’é™¤ blast æ¯”å¯¹ç»“æœä¸­çš„å†—ä½™é¡¹](#34-%e6%8e%92%e9%99%a4-blast-%e6%af%94%e5%af%b9%e7%bb%93%e6%9e%9c%e4%b8%ad%e7%9a%84%e5%86%97%e4%bd%99%e9%a1%b9)
  - [4 åŒæºæœç´¢ç»“æœçš„è¯„ä¼°](#4-%e5%90%8c%e6%ba%90%e6%90%9c%e7%b4%a2%e7%bb%93%e6%9e%9c%e7%9a%84%e8%af%84%e4%bc%b0)
- [å®éªŒé¡¹ç›®4: åŸºå› ç»„æ³¨é‡Šä¹‹ä»å¤´é¢„æµ‹ä¸ç»“æ„å»ºæ¨¡](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae4-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b3%a8%e9%87%8a%e4%b9%8b%e4%bb%8e%e5%a4%b4%e9%a2%84%e6%b5%8b%e4%b8%8e%e7%bb%93%e6%9e%84%e5%bb%ba%e6%a8%a1)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84-3)
  - [å®éªŒæµç¨‹](#%e5%ae%9e%e9%aa%8c%e6%b5%81%e7%a8%8b-3)
  - [1. åŸºå› ç»„æ•°æ®å‡†å¤‡](#1-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%95%b0%e6%8d%ae%e5%87%86%e5%a4%87)
  - [2. ä»å¤´åŸºå› é¢„æµ‹è½¯ä»¶çš„å®‰è£…ä¸æµ‹è¯•](#2-%e4%bb%8e%e5%a4%b4%e5%9f%ba%e5%9b%a0%e9%a2%84%e6%b5%8b%e8%bd%af%e4%bb%b6%e7%9a%84%e5%ae%89%e8%a3%85%e4%b8%8e%e6%b5%8b%e8%af%95)
  - [3. å…¨åŸºå› ç»„çš„ä»å¤´åŸºå› é¢„æµ‹](#3-%e5%85%a8%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%9a%84%e4%bb%8e%e5%a4%b4%e5%9f%ba%e5%9b%a0%e9%a2%84%e6%b5%8b)
    - [3.1 ä»»é€‰ä¸€ä¸ªèƒ½å¤Ÿè¿›è¡Œä»å¤´é¢„æµ‹åŸºå› çš„è½¯ä»¶ï¼Œå¦‚ Augustus. GeneMarkES/ET ç­‰](#31-%e4%bb%bb%e9%80%89%e4%b8%80%e4%b8%aa%e8%83%bd%e5%a4%9f%e8%bf%9b%e8%a1%8c%e4%bb%8e%e5%a4%b4%e9%a2%84%e6%b5%8b%e5%9f%ba%e5%9b%a0%e7%9a%84%e8%bd%af%e4%bb%b6%e5%a6%82-augustus-genemarkeset-%e7%ad%89)
    - [3.2 ä½¿ç”¨è¯¥è½¯ä»¶å¯¹ç¬¬ 1 æ­¥å‡†å¤‡çš„åŸºå› ç»„åºåˆ—è¿›è¡ŒåŸºå› é¢„æµ‹åˆ†æï¼Œ](#32-%e4%bd%bf%e7%94%a8%e8%af%a5%e8%bd%af%e4%bb%b6%e5%af%b9%e7%ac%ac-1-%e6%ad%a5%e5%87%86%e5%a4%87%e7%9a%84%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%ba%8f%e5%88%97%e8%bf%9b%e8%a1%8c%e5%9f%ba%e5%9b%a0%e9%a2%84%e6%b5%8b%e5%88%86%e6%9e%90)
  - [4 ä»å¤´åŸºå› é¢„æµ‹ç»“æœçš„é‰´åˆ«](#4-%e4%bb%8e%e5%a4%b4%e5%9f%ba%e5%9b%a0%e9%a2%84%e6%b5%8b%e7%bb%93%e6%9e%9c%e7%9a%84%e9%89%b4%e5%88%ab)
    - [4.1 å·²çŸ¥è›‹ç™½åºåˆ—](#41-%e5%b7%b2%e7%9f%a5%e8%9b%8b%e7%99%bd%e5%ba%8f%e5%88%97)
    - [4.2 åˆ›å»ºæœ¬åœ° BLAST æ•°æ®åº“](#42-%e5%88%9b%e5%bb%ba%e6%9c%ac%e5%9c%b0-blast-%e6%95%b0%e6%8d%ae%e5%ba%93)
    - [4.3 ä»GFFæ–‡æ¡£ä¸­æå–FASTAåºåˆ—](#43-%e4%bb%8egff%e6%96%87%e6%a1%a3%e4%b8%ad%e6%8f%90%e5%8f%96fasta%e5%ba%8f%e5%88%97)
      - [GFFä¸­åºåˆ—æ ¼å¼èŒƒä¾‹](#gff%e4%b8%ad%e5%ba%8f%e5%88%97%e6%a0%bc%e5%bc%8f%e8%8c%83%e4%be%8b)
      - [æå–åºåˆ—å‡½æ•°](#%e6%8f%90%e5%8f%96%e5%ba%8f%e5%88%97%e5%87%bd%e6%95%b0)
    - [4.4 ä½¿ç”¨åˆé€‚çš„ blast ç¨‹åºå¯¹è¯¥é¢„æµ‹åŸºå› ä¸å·²çŸ¥è›‹ç™½åºåˆ—è¿›è¡Œæ¯”å¯¹,ä»¥æ­¤æ¥é‰´åˆ«ä»å¤´é¢„æµ‹å‡ºæ¥çš„åŸºå› ](#44-%e4%bd%bf%e7%94%a8%e5%90%88%e9%80%82%e7%9a%84-blast-%e7%a8%8b%e5%ba%8f%e5%af%b9%e8%af%a5%e9%a2%84%e6%b5%8b%e5%9f%ba%e5%9b%a0%e4%b8%8e%e5%b7%b2%e7%9f%a5%e8%9b%8b%e7%99%bd%e5%ba%8f%e5%88%97%e8%bf%9b%e8%a1%8c%e6%af%94%e5%af%b9%e4%bb%a5%e6%ad%a4%e6%9d%a5%e9%89%b4%e5%88%ab%e4%bb%8e%e5%a4%b4%e9%a2%84%e6%b5%8b%e5%87%ba%e6%9d%a5%e7%9a%84%e5%9f%ba%e5%9b%a0)
    - [4.5 æŠŠ 4.4 ç»“æœåˆå¹¶åˆ° 3.2 è·å¾—çš„ GFF æ ¼å¼ç»“æœä¸­](#45-%e6%8a%8a-44-%e7%bb%93%e6%9e%9c%e5%90%88%e5%b9%b6%e5%88%b0-32-%e8%8e%b7%e5%be%97%e7%9a%84-gff-%e6%a0%bc%e5%bc%8f%e7%bb%93%e6%9e%9c%e4%b8%ad)
      - [ä¿¡æ¯æ•´åˆè„šæœ¬](#%e4%bf%a1%e6%81%af%e6%95%b4%e5%90%88%e8%84%9a%e6%9c%ac)
  - [5. ä»å¤´é¢„æµ‹ç»“æœçš„è¯„ä¼°](#5-%e4%bb%8e%e5%a4%b4%e9%a2%84%e6%b5%8b%e7%bb%93%e6%9e%9c%e7%9a%84%e8%af%84%e4%bc%b0)
    - [5.1 gffcompareå¯¹æ¯”](#51-gffcompare%e5%af%b9%e6%af%94)
    - [5.2 gffcompareç»“æœè§£æ](#52-gffcompare%e7%bb%93%e6%9e%9c%e8%a7%a3%e6%9e%90)
    - [5.3 ä¸å®éªŒä¸‰/åŒæºé¢„æµ‹ç»“æœè¿›è¡Œå¯¹æ¯”](#53-%e4%b8%8e%e5%ae%9e%e9%aa%8c%e4%b8%89%e5%90%8c%e6%ba%90%e9%a2%84%e6%b5%8b%e7%bb%93%e6%9e%9c%e8%bf%9b%e8%a1%8c%e5%af%b9%e6%af%94)
- [å®éªŒé¡¹ç›®5: åŸºå› ç»„æ³¨é‡Šä¹‹å¯åŠ¨å­åˆ†æå’Œé¢„æµ‹](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae5-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b3%a8%e9%87%8a%e4%b9%8b%e5%90%af%e5%8a%a8%e5%ad%90%e5%88%86%e6%9e%90%e5%92%8c%e9%a2%84%e6%b5%8b)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84-4)
  - [å®éªŒæµç¨‹](#%e5%ae%9e%e9%aa%8c%e6%b5%81%e7%a8%8b-4)
  - [1. æ•°æ®å‡†å¤‡](#1-%e6%95%b0%e6%8d%ae%e5%87%86%e5%a4%87-1)
  - [2. å¯åŠ¨å­ç›¸å…³DNAå…ƒä»¶HMMæ•°æ®](#2-%e5%90%af%e5%8a%a8%e5%ad%90%e7%9b%b8%e5%85%b3dna%e5%85%83%e4%bb%b6hmm%e6%95%b0%e6%8d%ae)
    - [2.1 ä»EPDæ•°æ®åº“ä¸­ä¸‹è½½ä»»æ„ä¸€ç§å¯åŠ¨å­ç›¸å…³çš„DNAå…ƒä»¶çš„HMMæ•°æ®](#21-%e4%bb%8eepd%e6%95%b0%e6%8d%ae%e5%ba%93%e4%b8%ad%e4%b8%8b%e8%bd%bd%e4%bb%bb%e6%84%8f%e4%b8%80%e7%a7%8d%e5%90%af%e5%8a%a8%e5%ad%90%e7%9b%b8%e5%85%b3%e7%9a%84dna%e5%85%83%e4%bb%b6%e7%9a%84hmm%e6%95%b0%e6%8d%ae)
  - [3 DNAå…ƒä»¶çš„è®¡ç®—é‰´åˆ«](#3-dna%e5%85%83%e4%bb%b6%e7%9a%84%e8%ae%a1%e7%ae%97%e9%89%b4%e5%88%ab)
    - [3.1 æ ¹æ®è¯¥ HMM æ•°æ®ï¼Œç¼–å†™ç¨‹åºå¯¹ä¸Šè¿°åŸºå› ç»„åºåˆ—è¿›è¡Œéå†ï¼Œè®¡ç®—åŸå§‹å¾—åˆ†](#31-%e6%a0%b9%e6%8d%ae%e8%af%a5-hmm-%e6%95%b0%e6%8d%ae%e7%bc%96%e5%86%99%e7%a8%8b%e5%ba%8f%e5%af%b9%e4%b8%8a%e8%bf%b0%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%ba%8f%e5%88%97%e8%bf%9b%e8%a1%8c%e9%81%8d%e5%8e%86%e8%ae%a1%e7%ae%97%e5%8e%9f%e5%a7%8b%e5%be%97%e5%88%86)
    - [3.2 åœ¨è®¡ç®—åˆ†å€¼çš„åŒæ—¶ï¼Œä½¿ç”¨ bootstrap æŠ½æ ·è¯„ä¼°çš„æ–¹æ³•å¯¹è®¡ç®—æ¯ä¸ªç‰‡æ®µå¯é æ€§på€¼](#32-%e5%9c%a8%e8%ae%a1%e7%ae%97%e5%88%86%e5%80%bc%e7%9a%84%e5%90%8c%e6%97%b6%e4%bd%bf%e7%94%a8-bootstrap-%e6%8a%bd%e6%a0%b7%e8%af%84%e4%bc%b0%e7%9a%84%e6%96%b9%e6%b3%95%e5%af%b9%e8%ae%a1%e7%ae%97%e6%af%8f%e4%b8%aa%e7%89%87%e6%ae%b5%e5%8f%af%e9%9d%a0%e6%80%a7p%e5%80%bc)
    - [3.3 æ ¹æ® p å€¼å¤§å°è¿›è¡Œè¿‡æ»¤](#33-%e6%a0%b9%e6%8d%ae-p-%e5%80%bc%e5%a4%a7%e5%b0%8f%e8%bf%9b%e8%a1%8c%e8%bf%87%e6%bb%a4)
  - [4. æŠŠåˆ†æç»“æœä¸åŸºå› ç»„çš„æ³¨é‡Šä¿¡æ¯è¿›è¡Œå¯¹æ¯”](#4-%e6%8a%8a%e5%88%86%e6%9e%90%e7%bb%93%e6%9e%9c%e4%b8%8e%e5%9f%ba%e5%9b%a0%e7%bb%84%e7%9a%84%e6%b3%a8%e9%87%8a%e4%bf%a1%e6%81%af%e8%bf%9b%e8%a1%8c%e5%af%b9%e6%af%94)
    - [4.1 è¿è¡Œ3ä¸­ç¼–å†™çš„ä»£ç ï¼Œå¾—åˆ°åˆ†æç»“æœ](#41-%e8%bf%90%e8%a1%8c3%e4%b8%ad%e7%bc%96%e5%86%99%e7%9a%84%e4%bb%a3%e7%a0%81%e5%be%97%e5%88%b0%e5%88%86%e6%9e%90%e7%bb%93%e6%9e%9c)
    - [4.2 ç¼–å†™ä»£ç è¿›è¡Œä½ç‚¹å¯è§†åŒ–](#42-%e7%bc%96%e5%86%99%e4%bb%a3%e7%a0%81%e8%bf%9b%e8%a1%8c%e4%bd%8d%e7%82%b9%e5%8f%af%e8%a7%86%e5%8c%96)
      - [æ­£é“¾](#%e6%ad%a3%e9%93%be)
      - [è´Ÿé“¾](#%e8%b4%9f%e9%93%be)
      - [é¢„æµ‹å…ƒä»¶æ•°ä¸åŸºå› æ•°çš„ç»Ÿè®¡](#%e9%a2%84%e6%b5%8b%e5%85%83%e4%bb%b6%e6%95%b0%e4%b8%8e%e5%9f%ba%e5%9b%a0%e6%95%b0%e7%9a%84%e7%bb%9f%e8%ae%a1)
    - [4.3 HMMåˆ†ç±»å™¨æ•ˆæœè¯„ä¼°](#43-hmm%e5%88%86%e7%b1%bb%e5%99%a8%e6%95%88%e6%9e%9c%e8%af%84%e4%bc%b0)
      - [æ¥è‡ªYJM1342çš„gffæ–‡æ¡£](#%e6%9d%a5%e8%87%aayjm1342%e7%9a%84gff%e6%96%87%e6%a1%a3)
      - [è¿›ä¸€æ­¥æ¢ç©¶](#%e8%bf%9b%e4%b8%80%e6%ad%a5%e6%8e%a2%e7%a9%b6)
    - [4.4 æ ¹æ®ä¸Šä¸€æ­¥è®¡ç®—çš„é˜ˆå€¼ï¼Œå¯¹ç¬¬3æ­¥çš„ç»“æœè¿›è¡Œè¿›ä¸€æ­¥çš„ç­›é€‰ï¼Œå¹¶æŒ‰ç…§ GFF3 æ ¼å¼ä¿å­˜](#44-%e6%a0%b9%e6%8d%ae%e4%b8%8a%e4%b8%80%e6%ad%a5%e8%ae%a1%e7%ae%97%e7%9a%84%e9%98%88%e5%80%bc%e5%af%b9%e7%ac%ac3%e6%ad%a5%e7%9a%84%e7%bb%93%e6%9e%9c%e8%bf%9b%e8%a1%8c%e8%bf%9b%e4%b8%80%e6%ad%a5%e7%9a%84%e7%ad%9b%e9%80%89%e5%b9%b6%e6%8c%89%e7%85%a7-gff3-%e6%a0%bc%e5%bc%8f%e4%bf%9d%e5%ad%98)
- [å®éªŒé¡¹ç›®6: åŸºå› ç»„å¯è§†åŒ–](#%e5%ae%9e%e9%aa%8c%e9%a1%b9%e7%9b%ae6-%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%8f%af%e8%a7%86%e5%8c%96)
  - [å®éªŒç›®çš„](#%e5%ae%9e%e9%aa%8c%e7%9b%ae%e7%9a%84-5)
  - [1. åŸºå› ç»„æµ‹åºå¯è§†åŒ–å·¥å…·ç›¸å…³æ–‡çŒ®èµ„æ–™çš„è°ƒç ”](#1-%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%b5%8b%e5%ba%8f%e5%8f%af%e8%a7%86%e5%8c%96%e5%b7%a5%e5%85%b7%e7%9b%b8%e5%85%b3%e6%96%87%e7%8c%ae%e8%b5%84%e6%96%99%e7%9a%84%e8%b0%83%e7%a0%94)
    - [1.1 Search in Pubmed](#11-search-in-pubmed)
    - [1.2 USCS Genome Browser](#12-uscs-genome-browser)
  - [2. åŸºå› ç»„å¯è§†åŒ–](#2-%e5%9f%ba%e5%9b%a0%e7%bb%84%e5%8f%af%e8%a7%86%e5%8c%96)
    - [2.1 ä¸‹è½½IGV](#21-%e4%b8%8b%e8%bd%bdigv)
    - [2.2 è¿›è¡ŒåŸºå› ç»„æœ¬åœ°å¯è§†åŒ–](#22-%e8%bf%9b%e8%a1%8c%e5%9f%ba%e5%9b%a0%e7%bb%84%e6%9c%ac%e5%9c%b0%e5%8f%af%e8%a7%86%e5%8c%96)
      - [2.2.1 å¯¼å…¥åŸºå› ç»„](#221-%e5%af%bc%e5%85%a5%e5%9f%ba%e5%9b%a0%e7%bb%84)
      - [2.2.2 å¤„ç†åŸGFFæ³¨é‡Šæ–‡ä»¶](#222-%e5%a4%84%e7%90%86%e5%8e%9fgff%e6%b3%a8%e9%87%8a%e6%96%87%e4%bb%b6)
      - [2.2.3 å¯¼å…¥GFFæ³¨é‡Šæ–‡ä»¶](#223-%e5%af%bc%e5%85%a5gff%e6%b3%a8%e9%87%8a%e6%96%87%e4%bb%b6)

## å®éªŒé¡¹ç›®1ï¼šåŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ

### å®éªŒç›®çš„

1. åŠ æ·±å…¨åŸºå› ç»„é¸Ÿæªæ³•æµ‹åºåŸç†çš„ç†è§£
2. ç†Ÿæ‚‰å’ŒæŒæ¡å¸¸ç”¨åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿå’Œè¯„ä¼°ç¨‹åºçš„ä½¿ç”¨
3. èƒ½å¤Ÿç‹¬ç«‹åœ°ä½¿ç”¨è‡ªå·±æ‰€æŒæ¡ç»Ÿè®¡å­¦åˆ†ææ‰‹æ®µæ¥å¯¹å®éªŒæ•°æ®è¿›è¡Œå¯¹æ¯”åˆ†æ
4. åŠ å¼ºå­¦ç”Ÿè‡ªä¸»å­¦ä¹ èƒ½åŠ›
5. åŠ æ·±ç»Ÿè®¡å­¦åœ¨åŸºå› ç»„æ•°æ®åˆ†æä¸­çš„åº”ç”¨
6. åŸ¹å…»å­¦ç”Ÿå‘ç°é—®é¢˜, åˆ†æé—®é¢˜å’Œè§£å†³é—®é¢˜çš„èƒ½åŠ›ã€‚


### å®éªŒæµç¨‹

```viz
digraph flowchart_1 {
    # rankdir=LR;
    fontname="Courier New";
    size="6,4"; ratio = fill;
    node [style="filled,setlinewidth(3)", color="#8383cc", fontname="Courier New", shape="Mrecord",fixedsize=true,width=2.5,fillcolor="#d9e7ee"];
    edge [color="0.635 0.707 0.707", fontname="Courier New"];
    label="åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ";
    step1[label="art_illuminaè°ƒç ”"];
    step2[label="åŸºå› ç»„æ•°æ®ä¸‹è½½"];
    step3[label="åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ"];

    sub_step1->step1;
    step1->sub_step2;
    sub_step2->step2;
    step2->step3;
    step3->sub_step31;

    subgraph cluster_1 {
        style=filled;
        color=lightgrey;
        sub_step1[label="æ–‡çŒ®è°ƒç ”"];
        sub_step2[label="è½¯ä»¶çš„å®‰è£…å’Œæµ‹è¯•"];
        label="æ‹“å±•å†…å®¹";
        subgraph cluster_2{
            color=grey;
            sub_step31[label="åŸºå› ç»„æµ‹åºæ•°æ®ä¸‹è½½"];
            sub_step32[label="ç›¸å…³çš„ç»Ÿè®¡è®¡ç®—"];
            sub_step33[label="art_profile_illumina"];
            sub_step34[label="å¯¹æ¯”profile"];
            sub_step31->sub_step32->sub_step33->sub_step34;
            label="æ•°æ®æ¨¡å‹çš„åˆ›å»º";
        }
    }
}
```

### 1. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿå·¥å…·ç›¸å…³æ–‡çŒ®èµ„æ–™çš„è°ƒç ”

#### From `pubmed`

> There are a number of existing software packages available for generating synthetic NGS read data, each tending to specialize on a particular attribute of a dataset. For example, ART [2], CuReSim [3], GemSim [4], and pIRS [5] focus on realistically emulating the biases inherent in the base calling of various next-generation sequencing (NGS) platforms. [ğŸ’­](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5125660/ "Link")

__CureSim__

> ___Comparison of mapping algorithms used in high-throughput sequencing: application to Ion Torrent data___ SÃ©golÃ¨ne Caboche,corresponding author1,4 Christophe Audebert,2,4 Yves Lemoine,1,3,4 and David Hot1,3,4

Curesim(ä¸€ç§æ–°çš„è¯»å–æ¨¡æ‹Ÿå™¨)å¯ä»¥é€šè¿‡è°ƒæ•´é”™è¯¯ç±»å‹çš„å‚æ•°ï¼Œä¸ºä»»ä½•ç±»å‹çš„ __HTS(é«˜é€šé‡æµ‹åº)æŠ€æœ¯__ ç”Ÿæˆå®šåˆ¶çš„åŸºå‡†æ•°æ®ã€‚
å…¶æœ¬è´¨æ˜¯ä¸€ä¸ªåŸºå‡†ç¨‹åºæ¥æ¯”è¾ƒä½¿ç”¨çœŸå®å’Œæ¨¡æ‹Ÿæ•°æ®é›†çš„HTSä¸­çš„æ˜ å°„ç®—æ³•ï¼Œå¹¶è€ƒè™‘äº†å››ä¸ªè¯„ä¼°æ ‡å‡†:

* è®¡ç®—èµ„æºå’Œæ—¶é—´éœ€æ±‚
* æ˜ å°„çš„ç¨³å¥æ€§
* èƒ½å¤ŸæŠ¥å‘Šé‡å¤åŒºåŸŸçš„è¯»å–ä½ç½®
* å¹¶èƒ½å¤Ÿæ£€ç´¢çœŸå®çš„é—ä¼ å˜å¼‚ä½ç½®ã€‚

ä¸ºåº¦é‡ç¨³å¥æ€§ï¼Œå…¶å¼•å…¥äº†ä¸€ä¸ªæ–°çš„å®šä¹‰ï¼Œç”¨äºæ­£ç¡®æ˜ å°„çš„è¯»å–ï¼Œä¸ä»…è¦è€ƒè™‘è¯»å–çš„é¢„æœŸèµ·å§‹ä½ç½®ï¼Œè¿˜è¦è€ƒè™‘ç»“æŸä½ç½®ä»¥åŠç´¢å¼•å’Œæ›¿æ¢çš„æ•°é‡ã€‚

> ç”±äºåŸºå› ç»„é‡æ–°æµ‹åºï¼ˆå…¨åŸºå› ç»„æµ‹åºå’Œé¶å‘æµ‹åºï¼‰ç­‰ç‰¹å®šåº”ç”¨çš„å¼€å‘ï¼ŒHTSåœ¨é—ä¼ å­¦å’ŒåŸºå› ç»„å­¦æ–¹é¢äº§ç”Ÿäº†å¤§é‡çŸ¥è¯†ã€‚è¿™ç§æŠ€æœ¯æ¼”å˜ä¸å¼€å‘æ–°ç®—æ³•ä»¥å¤„ç†æ‰€äº§ç”Ÿçš„readsçš„æ•°é‡å’Œè´¨é‡å¹¶è¡Œã€‚é‡æ–°æµ‹åºæ–¹æ³•çš„åŸºæœ¬åˆ†ææ­¥éª¤æ˜¯å°†readsæ˜ å°„åˆ°å‚è€ƒåŸºå› ç»„ä¸Šã€‚è¯¥æ­¥éª¤æ¶‰åŠå°†è¯»æ•°å‡†ç¡®å®šä½åˆ°å‚è€ƒåŸºå› ç»„åºåˆ—ä¸Šï¼Œè¿™éå¸¸é‡è¦ï¼Œå› ä¸ºå®ƒå†³å®šäº†ä¸‹æ¸¸åˆ†æçš„å…¨å±€è´¨é‡ã€‚ç”¨äºæ­¤æ­¥éª¤çš„ç®—æ³•ç§°ä¸ºæ˜ å°„å™¨ã€‚æ˜ å°„å™¨å¿…é¡»æ•æ„Ÿä¸”å‡†ç¡®ï¼Œå¹¶ä¸”é€Ÿåº¦è¦å¿«ï¼Œä¸èƒ½è¦å¤ªå¤šçš„è®¡ç®—ã€‚ä»–ä»¬åº”è¯¥èƒ½å¤Ÿåœ¨å‚è€ƒåŸºå› ç»„ä¸Šæ‰¾åˆ°æ¯ä¸€ä¸ªè¯»æ•°çš„çœŸå®ä½ç½®ï¼Œå¹¶ç†æƒ³åœ°åŒºåˆ†æŠ€æœ¯æµ‹åºé”™è¯¯å’Œè‡ªç„¶é—ä¼ å˜å¼‚ã€‚[ğŸ’­](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4051166/ "Link")

### 2. æŸ¥é˜… art_illumina è½¯ä»¶ä¸­çš„å¸¸ç”¨æµ‹åºå¹³å°ä¿¡æ¯

[ğŸ’­](https://www.cnblogs.com/think-and-do/p/6638157.html "Link")

```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
------         2016/6/5     22:41         234519 Emp100R1.txt
------         2016/6/5     22:41         409657 EmpMiSeq250R1.txt
------         2016/6/5     22:41          77918 EmpR36R1.txt
------         2016/6/5     22:41         150841 HiSeq2500L125R1.txt
...
```

æŸ¥é˜…```art_illumina_README```å¯ä»¥çŸ¥é“ï¼Œé€šè¿‡æŸ¥çœ‹```\art_bin_MountRainier\Illumina_profiles```ç›®å½•ä¸‹çš„æ–‡ä»¶å¯ä»¥æŸ¥çœ‹art_illuminaæ”¯æŒçš„æµ‹åº

### 3. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿè½¯ä»¶çš„å®‰è£…å’Œé¢„æµ‹

__ARTè½¯ä»¶ä¸‹è½½__

* <https://www.niehs.nih.gov/research/resources/software/biostatistics/art/>
* <https://www.niehs.nih.gov/research/resources/assets/docs/artsrcmountrainier2016.06.05linux.tgz>

#### 3.1 Win10å¹³å°

<table align="center">
  <tr>
    <td>
      <img src="./figs/artè½¯ä»¶ä¸‹è½½.jpg"></img>
    </td>
    <td>
      <img src="./figs/artå‘½ä»¤è¡Œç•Œé¢_new.png"></img>
    </td>
  </tr>
  <tr>
    <td>artè½¯ä»¶ä¸‹è½½</td>
    <td>artå‘½ä»¤è¡Œç•Œé¢</td>
  </tr>
</table>

è€ƒè™‘åˆ°Windowså¹³å°çš„artç³»åˆ—è½¯ä»¶è¿è¡Œä¸ç¨³å®š, åŒæ—¶åœ¨Ubuntuä¸‹è¿›è¡Œå®éªŒã€‚

#### 3.2 Ubuntu

```bash
# cd åˆ°ç›®æ ‡å®‰è£…ç›®å½•
wget https://www.niehs.nih.gov/research/resources/assets/docs/artsrcmountrainier2016.06.05linux.tgz
tar -d artsrcmountrainier2016.06.05linux.tgz
cd art_src_MountRainier_Linux/
cat README
```

æŸ¥çœ‹å®‰è£…è¦æ±‚ï¼š

```README
...
COMPILATION AND INSTALLATION

	PREREQUISITES:

		1) GNU g++ 4.0 or above (http://gcc.gnu.org/install)
		2) GNU gsl library (http://www.gnu.org/s/gsl/)
...
```

è¡¥å®‰è£…G++ä¸GSLåº“:

```bash
# cd åˆ°é¡¶å±‚ç›®å½•
apt-get install g++
apt-get install libgsl0-dev
```

æ‰§è¡ŒARTç³»åˆ—çš„å®‰è£…

```bash
# cd åˆ°ç›®æ ‡å®‰è£…ç›®å½•
./configure --prefix=$HOME
make
make install
```

åˆ°æ­¤è½¯ä»¶å®‰è£…å®Œæˆã€‚

#### åŸç†

ä½¿ç”¨å†…ç½®çš„ï¼Œç‰¹å®šäºæŠ€æœ¯çš„è¯»å–é”™è¯¯æ¨¡å‹å’Œåœ¨å¤§å‹æµ‹åºæ•°æ®é›†ä¸­å‡­ç»éªŒå‚æ•°åŒ–çš„åŸºæœ¬è´¨é‡å€¼é…ç½®æ–‡ä»¶ï¼Œå³åˆ›å»ºProfileæ–‡æ¡£ï¼Œåˆ©ç”¨ç»éªŒåˆ†å¸ƒ(ä½ç‚¹. è´¨é‡åˆ†æ•°. é¢‘æ•°. ç¢±åŸºå››ç»´åº¦ä¿¡æ¯)ï¼Œæ¥æ¨¡æ‹Ÿæµ‹åºè¿‡ç¨‹ï¼Œä»è€Œç”Ÿæˆæ¨¡æ‹Ÿçš„æµ‹åºreadsã€‚

#### IO

ç”±art_illuminaçš„å‚æ•°è®¾ç½®å¯ä»¥çœ‹åˆ°ï¼Œå¯ä»¥è¿›è¡Œå¤šç§è¦†ç›–åº¦. å•åŒç«¯æµ‹åº. å¤šç§æµ‹åºå¹³å°ç­‰çš„æ¨¡æ‹Ÿæµ‹åºã€‚

åŒæ—¶ï¼Œä»¥åŸºå› ç»„æ–‡ä»¶ä¸ºè¾“å…¥æ–‡ä»¶ï¼Œè¾“å‡ºFASTQæ ¼å¼çš„åºåˆ—æ–‡æ¡£(å³åŒ…å«åºåˆ—è´¨é‡ä¿¡æ¯)ï¼Œä»¥åŠaln,sam,statæ ¼å¼çš„æ¯”å¯¹æ–‡æ¡£ã€‚

##### FASTQ

* æ¯æ¡è®°å½•(Read)æœ‰å››è¡Œä¿¡æ¯
* ç¬¬ä¸€è¡Œï¼šReadçš„åç§°ï¼Œå®éªŒæ‰‹æ®µï¼Œæµ‹åºå¹³å°ç­‰
* ç¬¬äºŒè¡Œï¼šSequence
* ç¬¬ä¸‰è¡Œï¼šignore
* ç¬¬å››è¡Œï¼šBase Quality
  * ASCIIç¼–ç çš„è´¨é‡åˆ†æ•°Q
  * $Q=-10\log_{10}p$
  * p: Probability of incorrect base call
  * Phred33: ASCII=chr(Q + 33)
  * i.e p=0.05 -> Q = 13.01.. -> ```'.'```



##### SAM

ä¸€ç§åºåˆ—æ¯”å¯¹æ ¼å¼æ ‡å‡†ï¼Œ ç”±sangeråˆ¶ å®šï¼Œæ˜¯ä»¥TABä¸ºåˆ†å‰²ç¬¦çš„æ–‡æœ¬æ ¼å¼ã€‚ä¸»è¦åº”ç”¨äºæµ‹åºåºåˆ— mappingåˆ°åŸºå› ç»„ä¸Š çš„ç»“æœè¡¨ç¤ºï¼Œè¡¨ç¤ºä»»æ„çš„å¤šé‡æ¯”å¯¹ç»“æœã€‚

* æ³¨é‡Šä¿¡æ¯éƒ¨åˆ† (```@```ç¬¦åˆå¼€å¤´)
* æ¯”å¯¹ç»“æœéƒ¨åˆ†

##### STAT

è¡¨å¾åŸºå› ç»„å„ä¸ªæŸ“è‰²ä½“ä¸Šæ¯ä¸ªä½ç‚¹çš„è¦†ç›–åº¦

* å‚è€ƒä½ç½®
* ä»è¯¥ä½ç½®èµ·å§‹çš„readsæ•°é‡
* è¦†ç›–åˆ°è¯¥ä½ç½®çš„readsæ•°é‡

ç†è®ºä¸Šæ¯ä¸ªä½ç‚¹å¯¹åº”èµ·å§‹çš„readæ•°é‡å¤§è‡´ä¸€è‡´ï¼Œä¸”æ•°é‡ä¸å¤šï¼›è€Œè¦†ç›–åˆ°è¯¥ä½ç‚¹çš„readsæ•°é‡ä¸å…¶ä½ç‚¹ä¹Ÿåº”æ— æ˜æ˜¾åŒºåˆ«ï¼Œä¸”æ•°é‡åº”æ¯”èµ·å§‹readå¤šã€‚

### 4. åŸºå› ç»„æ•°æ®ä¸‹è½½

#### 4.1 æœç´¢æ•°æ®		

key|value
-|-
Organism/Name|Saccharomyces cerevisiae YJM1342
Strain|YJM1342
BioSample|SAMN01923166
BioProject|PRJNA189900
Assembly|GCA_000977265.3
Size (Mb)|12.6226
GC%|38.48
WGS|\-
Scaffolds|17
Gene|6912
Protein|5409
Release Date|2015/02/24
Modify Date|2015/04/09

äºGenomeæ•°æ®åº“æœç´¢ï¼Œé€‰æ‹©èŒæ ª (èŒæ ªç¼–å· YJM1342)ï¼›èŒæ ªä¿¡æ¯ (æœ‰å¯¹åº”åŸºå› åŠè›‹ç™½)å¦‚ä¸Šè¡¨ã€‚

#### 4.2 åŸºå› ç»„ç»„è£…åºåˆ—æ•°æ®ä¸‹è½½

ä¸‹è½½é“¾æ¥

```bash
# YJM1342 åŸºå› ç»„åºåˆ—æ–‡æ¡£
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/977/265/GCA_000977265.3_Sc_YJM1342_v1/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna.gz
gzip -d GCA_000977265.3_Sc_YJM1342_v1_genomic.fna.gz
# YJM1342 åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£
wget ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCA/000/977/265/GCA_000977265.3_Sc_YJM1342_v1/GCA_000977265.3_Sc_YJM1342_v1_genomic.gff.gz
gzip -d GCA_000977265.3_Sc_YJM1342_v1_genomic.gff.gz
```

åºåˆ—æ–‡æ¡£åœ¨æ¥ä¸‹æ¥çš„æ¨¡æ‹Ÿåºåˆ—ç»„è£…ä¸­ç”¨åˆ°ï¼Œæ³¨é‡Šæ–‡æ¡£å¾…ä¸‹æ¬¡å®éªŒåˆ©ç”¨ã€‚

æ–‡æ¡£åç§°ï¼š

```bash
GCA_000977265.3_Sc_YJM1342_v1_genomic.fna
GCA_000977265.3_Sc_YJM1342_v1_genomic.gff
```

#### 4.3 åŸºå› ç»„æµ‹åºæ•°æ®ä¸‹è½½[æ‹“å±•å†…å®¹]

ä¸ºç”Ÿæˆç‹¬ç«‹Profile,å› æ­¤éœ€è¦ä¸‹è½½æµ‹åºæ•°æ®ã€‚

<table align="center">
  <tr>
    <td>
      <img src="./figs/èŒæ ªSRAæµ‹åºä¿¡æ¯.png", width="50%"></img>
    </td>
  </tr>
  <tr>
    <td>èŒæ ªSRAæµ‹åºä¿¡æ¯</td>
  </tr>
</table>

##### 4.3.1 ä¿¡æ¯

* Instrument: Illumina HiSeq 2000
* Strategy: WGS
* Source: GENOMIC
* Selection: RANDOM
* Layout: __PAIRED__

å¯ä»¥çœ‹åˆ°è¯¥æµ‹åºæ•°æ®ä¸ºåŒæœ«ç«¯æµ‹åºè€Œäº§ç”Ÿ,åœ¨HiSeq 2000å¹³å°ä¸‹å®Œæˆæµ‹åºã€‚

##### 4.3.2 åˆ©ç”¨NCBI SRA Toolkitä¸‹è½½å’Œæå–è¯¥æµ‹åºæ•°æ®

```bash
# ä¸‹è½½ sra-toolkit
apt install sra-toolkit
# cdåˆ°ç›®æ ‡è·¯å¾„
wget https://sra-downloadb.be-md.ncbi.nlm.nih.gov/sos/sra-pub-run-2/SRR800817/SRR800817.1

# æˆ–
fastq-dump --split-3 SRR800817
# å¾—æ–‡ä»¶ï¼š
 SRR800817.1
 SRR800817_1.fastq
 SRR800817_2.fastq
```

##### 4.3.3 åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿçš„æ•°æ®æ¨¡å‹(profile)åˆ›å»º

```bash
../art_src_MountRainier_Linux/ART_profiler_illumina/art_profiler_illumina HiSeq2k_YJM1342 SRA_files fastq 20
```

æŸ¥çœ‹æ¨¡æ‹Ÿè¿ç®—çš„è¾“å‡ºç»“æœæ–‡ä»¶:

```bash
head -n 10 HiSeq2k_YJM1342R1.txt
```

```txt
.	0	3	8	11	17	19	21	23	24	26	27	28	29	30	31	32	34	35
.	0	8	9	11	19	21	22	24	26	34	50	82	111	113	189	625	712	1299
.	1	3	8	11	13	17	20	23	24	25	26	27	28	29	30	31	32	34	35
.	1	27	28	34	36	45	46	49	56	57	62	82	101	122	124	173	536	581	1299
.	2	3	8	11	16	17	20	21	23	24	26	27	28	29	30	31	32	34	35
.	2	9	10	14	24	34	37	38	41	44	53	76	106	125	128	173	514	567	1299
.	3	3	8	11	13	17	18	20	23	24	25	26	27	28	29	31	32	33	34	36	38
.	3	8	9	14	15	17	21	24	30	33	34	35	40	45	59	73	81	122	153	403	1299
.	4	3	8	11	17	18	20	21	23	24	26	27	28	29	31	32	33	34	36	38
.	4	21	22	30	36	42	44	46	51	53	57	66	69	76	86	87	123	158	413	1299
...
```

è¯´æ˜ProfileæˆåŠŸç”Ÿæˆã€‚ä¸‹é¢ä¸ illumina Read Profile è¿›è¡Œå¯¹æ¯”ï¼š

```bash
python3 visualize_HiSeq2k_Profile.py -i ./Illumina_profiles/
```

* è¾“å…¥æ–‡ä»¶
  * HiSeq2kL100R1.txt
  * HiSeq2kL100R2.txt
  * HiSeq2k_YJM1342R1.txt
  * HiSeq2k_YJM1342R2.txt
* è¾“å‡º

<table align="center">
  <tr>
    <td>
      <img src="./figs/selfProfile_è´¨é‡åˆ†æ•°.png"></img>
    </td>
    <td>
      <img src="./figs/selfProfile_é¢‘æ•°.png"></img>
    </td>
    <td>
      <img src="./figs/selfProfile_è´¨é‡åˆ†æ•°2.png"></img>
    </td>
    <td>
      <img src="./figs/selfProfile_é¢‘æ•°2.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342R1 è´¨é‡åˆ†æ•°</td>
    <td>YJM1342R1 é¢‘æ•°</td>
    <td>YJM1342R2 è´¨é‡åˆ†æ•°</td>
    <td>YJM1342R2 é¢‘æ•°</td>
  </tr>
  <tr>
    <td><img src="./figs/illuminaProfile_è´¨é‡åˆ†æ•°.png"></img></td>
    <td><img src="./figs/illuminaProfile_é¢‘æ•°.png"></img></td>
    <td><img src="./figs/illuminaProfile_è´¨é‡åˆ†æ•°2.png"></img></td>
    <td><img src="./figs/illuminaProfile_é¢‘æ•°2.png"></img></td>
  </tr>
  <tr>
    <td>HiSeq2kL100R1 è´¨é‡åˆ†æ•°</td>
    <td>HiSeq2kL100R1 é¢‘æ•°</td>
    <td>HiSeq2kL100R2 è´¨é‡åˆ†æ•°</td>
    <td>HiSeq2kL100R2 é¢‘æ•°</td>
  </tr>
</table>

é€”ä¸­åˆ—çš„ä¸åŒé¢œè‰²è¡¨ç¤ºä¸åŒç¢±åŸºï¼Œæ¨ªè½´ä¸ºä½ç‚¹ï¼Œçƒ­å›¾å±•ç¤ºé¢‘æ•°æˆ–è´¨é‡åˆ†æ•°ã€‚å¯ä»¥çœ‹åˆ°äºŒè€…çš„è´¨é‡åˆ†æ•°ä»¥åŠé¢‘æ•°åˆ†å¸ƒæœ‰æ‰€ä¸åŒï¼Œéœ€è¿›è¡Œä¸‹ä¸€æ­¥çš„ç»Ÿè®¡æ€§æ¯”è¾ƒã€‚

ç”±profileçš„ä½ç‚¹. è´¨é‡åˆ†æ•°. é¢‘æ•°. ç¢±åŸºå››ä¸ªç»´åº¦çš„æ•°æ®è¿›è¡Œè®¾è®¡ï¼š

* æŠŠä¸åŒç¢±åŸºæ•°æ®åˆ†ç¦»ï¼Œæ¯ç»„æ•°æ®åªå…·æœ‰ä½ç‚¹. è´¨é‡åˆ†æ•°. é¢‘æ•°ä¸‰ç»´ä¿¡æ¯
* å°†æ¯ç»„ä¸­çš„æ¯ä¸ªä½ç‚¹å¯¹åº”è´¨é‡åˆ†æ•°çš„é¢‘æ•°æ•°æ®è½¬åŒ–ä¸ºæ¦‚ç‡å½¢å¼(é¢‘æ•°/è¯¥ä½ç‚¹æ‰€æœ‰è´¨é‡åˆ†æ•°çš„æ€»é¢‘æ•°)
* ä¸Šä¸€æ­¥å³å¾—åˆ°äº†ä¸€æ¦‚ç‡çŸ©é˜µï¼Œå¯æ£€éªŒæ˜¯å¦ç¬¦åˆæ­£æ€åˆ†å¸ƒï¼Œè¿›è€Œè¿›è¡Œå¤šé‡æ£€éªŒæˆ–æ–¹å·®åˆ†æï¼Œå¦åˆ™è¿›è¡ŒKruskal-Wallisç§©å’Œæ£€éªŒ(ä¸¤æ ·æœ¬çš„Wilcoxonæ–¹æ³•åœ¨å¤šäºä¸¤æ ·æœ¬æ—¶çš„æ¨å¹¿)ï¼Œpaired(é…ä¼ç»„)è®¾è®¡ä¸­ï¼Œå¤šä¸ªæ ·æœ¬çš„æ¯”è¾ƒï¼Œè‹¥å®ƒä»¬çš„æ€»ä½“ä¸èƒ½æ»¡è¶³æ­£æ€æ€§å’Œæ–¹å·®é½æ¬¡æ€§çš„è¦æ±‚ï¼Œå¯é‡‡ç”¨Friedmanç§©å’Œæ£€éªŒ

### 5. åŸºå› ç»„æµ‹åºæ¨¡æ‹Ÿ

#### 5.1 ä½¿ç”¨artç³»åˆ—è½¯ä»¶, å¯¹ä¸‹è½½åŸºå› ç»„åºåˆ—è¿›è¡Œå…¨åŸºå› ç»„æµ‹åºè¿›è¡ŒåŒæœ«ç«¯æ¨¡æ‹Ÿ

é¦–å…ˆè¿›è¡Œè¦†ç›–åº¦ä¸º10çš„åˆæ­¥è¿è¡Œï¼Œåˆ¤æ–­è½¯ä»¶æ˜¯å¦å¯ä»¥æ­£å¸¸è¿è¡Œï¼Œç¤ºä¾‹è„šæœ¬ï¼š

```bash
# paired-end read simulation for HiSeq2000
./art_illumina -ss HS20 -sam -i ../../Genomics/data/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna -p -l 100 -f 10 -m 200 -s 10 -o ../../Genomics/data/YJM1342_0918_test
```

å‚æ•°è§£æ:

* -ss HS20 [æµ‹åºå¹³å°]
* -sam [è¾“å‡ºsamæ–‡ä»¶]
* -i .fna [è¾“å…¥æ–‡ä»¶ï¼Œå³è¾“å…¥çš„å‚è€ƒåŸºå› ç»„]
* -p [paired-endæµ‹åºï¼ŒçŸ­æ’å…¥ç‰‡æ®µ]
* -l 100 [read lengethï¼Œæµ‹åºé•¿åº¦]
* -f 10 [è¦†ç›–åº¦]
* -m 200 [paired-endçš„ç‰‡æ®µå¤§å°]
* -s 10 [-mç‰‡æ®µçš„åå·®]
* -o ../../Genomics/data/YJM1342_0918_test [è¾“å‡ºæ–‡ä»¶åå‰ç¼€]

æ­¤æ—¶ä¼šå‡ºç°é—®é¢˜ï¼Œwindowså¹³å°è¿è¡Œ```art_illumina```æ—¶ä¼šå‡ºç°```Error: the number of bases is not equal to the number of quality scores!```çš„æŠ¥é”™å†…å®¹ï¼Œä¸”æ— è®ºæ›´æ”¹å‚æ•°ä¸å¦ï¼Œè¾“å‡ºå†…å®¹éƒ½ä¸å¤Ÿç¨³å®šã€‚å› è€Œæ¥ä¸‹æ¥çš„å®éªŒéƒ½åœ¨Ubuntuå¹³å°ä¸‹è¿›è¡Œã€‚
<br />

#### 5.2 ä½¿ç”¨åˆé€‚çš„è½¯ä»¶æˆ–å·¥å…·æ¥æŸ¥çœ‹æ¨¡æ‹Ÿè¿ç®—çš„è¾“å‡ºç»“æœæ–‡ä»¶åŠå…¶å†…å®¹ã€‚

```bash
head -n 6 ../../Genomics/data/YJM1342_0918_test1.fq
```

```txt
@CP004425.2-20760/1
TAACTGGAAGGAAAAGAACAGATAAATGTCTCAAACAAAGCTGATCAAGCCGCGGTATTTATATGAAACTTTGAACAACTACATCTGCACACATGGGCTC
+
@@CF<FFFHHHGHJGIIIJHIH(?IIIHJGJAJIIJJHJJJHCH@HJIJIJG;'GHH>)FFIHAD=(DFGADBD.AH(+GD+CD+DCDFC;C@DDDDD+C
@CP004425.2-20758/1
CTAATGGAACCAGATCCATTCACCCATAAACGAGAAAATGGTTTGCCCAGTGGAACTTTGACAGCAGACTTCCTTGGTGTATTCAATTTTGTCTGAGAAT
```
å¯¹éƒ¨åˆ†fastaqæ•°æ®è¿›è¡Œç»Ÿè®¡ï¼Œå¾—å‡ºæ¯æ¡Readsæ¯ä¸ªç¢±åŸºçš„Qå€¼ï¼Œç»˜å‡ºç®±å‹å›¾ã€‚
<table align="center">
  <tr>
    <td>
      <img src="./figs/qc.png"></img>
    </td>
  </tr>
  <tr>
    <td>æ¯æ¡Readsæ¯ä¸ªç¢±åŸºçš„Qå€¼ç®±å‹å›¾</td>
  </tr>
</table>
<br />

å¯ä»¥çœ‹åˆ°é¦–å°¾è´¨é‡ç¨å·®ï¼Œä¸­é—´è¾ƒå¥½ã€‚

#### 5.3 ç»Ÿè®¡ä¸åŒå‚æ•°è®¾ç½®ä¸‹çš„æ¨¡æ‹Ÿç»“æœ

##### 5.3.1 SRR800817æµ‹åºè¦†ç›–åº¦ç†è®ºå€¼

å°†YJM1342çš„æµ‹åºå¤§å°é™¤ä»¥å…¶åŸºå› ç»„å¤§å°å³å¯çŸ¥è¯¥SRR800779æµ‹åºçš„ç†è®ºè¦†ç›–åº¦ç­‰ã€‚
|$m$|$e^{-m}$|$1-e^{-m}$|
:-:|:-:|:-:|:-:
|$7G/12.6226M \approx 554.56$|$1.44*10^{-241}$|$\approx 1$|

##### 5.3.2 ä¸åŒå‚æ•°è®¾ç½®ä¸‹è¦†ç›–åº¦æ•°æ®

è·å–åˆ°çš„ YJM1342 åŸºå› ç»„å¤§å°:12.6226Mb (Mb: $10^6$ bases)ã€‚art_illumina å¯ä»¥å¯¹å„ä¸ªè¦†ç›–åº¦è¿›è¡Œæ¨¡æ‹Ÿï¼Œéœ€è¦æµ‹è¯•çœ‹çœ‹å…·ä½“è¾“å‡ºç»“æœçš„æƒ…å†µ:

|æ¨¡æ‹Ÿçš„readsé•¿åº¦|pairedæµ‹åºæ¨¡æ‹Ÿseq-segå¹³å‡é•¿åº¦|sdev|è¦†ç›–åº¦/å•å€ä½“åŸºå› ç»„æ•° ($m$)|ç†è®ºä¸¢å¤±ç‡ ($P_0 = e^{-m}$)|è¦†ç›–ç‡($1-e^{-m}$)|å®é™…è¦†ç›–ç‡ï¼ˆæ ¹æ®fqæ–‡ä»¶åºåˆ—è¡Œ/åŸºå› ç»„å¤§å°å¾—æ¥çš„è¦†ç›–åº¦è®¡ç®—ï¼‰|
:-:|:-:|:-:|:-:|:-:|:-:|:-:
|100|200|10|1|0.37|0.63|0.39|
|100|200|10|2|0.14|0.86|0.63|
|100|200|10|3|0.05|0.95|0.77|
|100|200|10|4|0.02|0.98|0.86|
|100|200|10|5|0.01|0.99|0.91|

åˆ†åˆ«è®¾ç½®è¦†ç›–åº¦å‚æ•°ï¼Œå¯ä»¥çœ‹åˆ°ç†è®ºè¦†ç›–ç‡éšç€è¦†ç›–åº¦çš„å¢åŠ è€Œå¢åŠ ã€‚ä½†æ˜¯è¦†ç›–åº¦åˆ°äº†ä¸€å®šå€¼ï¼Œè¦†ç›–ç‡çš„å¢å¹…é€æ¸å‡å°,å¹¶ä¸”é€æ­¥é è¿‘1ã€‚è¦†ç›–åº¦çš„å†æé«˜å¯¹äºæå‡è¦†ç›–ç‡æ”¶æ•ˆç”šå¾®ã€‚
ä»å®é™…è¦†ç›–åº¦ä¸å¯¹åº”è¦†ç›–ç‡å¯ä»¥çœ‹åˆ°ï¼Œå®é™…è¦†ç›–åº¦ä¸ç†è®ºå­˜åœ¨å·®å€¼ï¼Œä½¿å¾—è¦†ç›–ç‡ä¸å¦‚ç†è®ºè¦†ç›–ç‡ã€‚

å°†ä¸Šè¿°è¡¨æ ¼æ•°æ®å¯è§†åŒ–å¦‚å›¾ï¼š

<table align="center">
  <tr>
    <td>
      <img src="./figs/è¦†ç›–åº¦è¦†ç›–ç‡_new.png"></img>
    </td>
  </tr>
  <tr>
    <td>è¦†ç›–ç‡y ~ è¦†ç›–åº¦xçš„æ›²çº¿å›¾</td>
  </tr>
</table>

ä¸‹é¢æ˜¯è¾“å‡ºæ–‡ä»¶å±•ç¤ºï¼š

```txt
# m = 1 -> [Paired-end sequencing simulation] Total CPU time used: 4.82569; Fold Coverage: 1X; The random seed for the run: 1568811948 ...
@CP004425.2-2076/1
GTATTTTTAATCAAGTGGAAAGATGAGTGGAAAAAAGGGCAATGAAATAGAAAAGGACAGGCCTGAAAGGGAAGAATACAAGAAGATTGAGTATATTGGA
...
# m = 2 -> [Paired-end sequencing simulation] Total CPU time used: 9.44064; Fold Coverage: 2X; The random seed for the run: 1568811977 ...
...
# m = 3 -> [Paired-end sequencing simulation] Total CPU time used: 14.4648; Fold Coverage: 3X; The random seed for the run: 1568812008 ...
...
# m = 4 -> [Paired-end sequencing simulation] Total CPU time used: 18.6218; Fold Coverage: 4X; The random seed for the run: 1568812057 ...
...
# m = 5 -> [Paired-end sequencing simulation] Total CPU time used: 23.4467; Fold Coverage: 5X; The random seed for the run: 1568812098 ...
...
```

## å®éªŒé¡¹ç›®2ï¼šåºåˆ—ç»„è£…

### å®éªŒç›®çš„

1. åŠ æ·±åºåˆ—ç»„è£…åŸç†çš„ç†è§£
2. ç†Ÿæ‚‰å¹¶æŒæ¡å¸¸ç”¨çš„åºåˆ—ç»„è£…è½¯ä»¶çš„ä½¿ç”¨ï¼ˆå¦‚ AllPathsLGã€ soapdenovo2ã€ Velvet ç­‰ï¼‰
3. åŠ å¼ºå­¦ç”Ÿè‡ªä¸»å­¦ä¹ èƒ½åŠ›
4. åŸ¹å…»å­¦ç”Ÿå‘ç°é—®é¢˜ã€åˆ†æé—®é¢˜å’Œè§£å†³é—®é¢˜çš„èƒ½åŠ›ã€‚


### å®éªŒæµç¨‹

```viz
digraph flowchart_2 {
    # rankdir=LR;
    fontname="Courier New";
    size="6,8"; ratio = fill;
    node [style="filled,setlinewidth(3)", color="#8383cc", fontname="Courier New", shape="Mrecord",fixedsize=true,width=2.5,fillcolor="#d9e7ee"];
    edge [color="0.635 0.707 0.707", fontname="Courier New"];
    label="åºåˆ—ç»„è£…";
    step1[label="æ•°æ®å‡†å¤‡"];
    step2[label="fastqcè´¨æ§åˆ†æ"];
    # step3[label="ä¸å‚è€ƒåŸºå› ç»„çš„æ¯”å¯¹"];
    # step4[label="åºåˆ—ç»„è£…åŠç»“æœåˆ†æ"];

subgraph cluster_1{
    style=filled;
    color=lightgrey;
    node [color=white];
    subgraph cluster_01 {
        color=grey;
        label="art_illumina æ¨¡æ‹ŸåŒæœ«ç«¯æµ‹åº"
        sub_c1_1[label="çŸ­æ’å…¥ç‰‡æ®µåº“"];
        sub_c1_2[label="é•¿æ’å…¥ç‰‡æ®µåº“"];
    }
    label="æ•°æ®å‡†å¤‡"
    sub_c1_3[label="å‚è€ƒåŸºå› ç»„æ–‡ä»¶åˆ›å»º"];
}
    sub_c1_1->sub_c1_3;
    sub_c1_2->sub_c1_3;

    subgraph cluster_0 {
        style=filled;
        color=lightgrey;
        sub_step1[label="æ–‡çŒ®è°ƒç ”"];
        sub_step2[label="è½¯ä»¶çš„å®‰è£…å’Œæµ‹è¯•"];
        sub_step3[label="åºåˆ—ç»„è£…ç»ƒä¹ "];
        label="æ‹“å±•å†…å®¹";
        sub_step1->sub_step2->sub_step3;
    }

    subgraph cluster_2 {
        style=filled;
        color=lightgrey;
        node [color=white];
        sub_c2_step1[label="bowtie2å¯¹æ¯”"];
        sub_c2_step2[label="samtoolsç»Ÿè®¡åˆ†æ"];
        sub_c2_step3[label="plot-bamstatså¯è§†åŒ–"];
        label="ä¸å‚è€ƒåŸºå› ç»„çš„æ¯”å¯¹";
        sub_c2_step1->sub_c2_step2->sub_c2_step3;
    }

    subgraph cluster_3 {
        style=filled;
        color=lightgrey;
        node [color=white];
        sub_c3_step1[label="SOAPdenovoç»„è£…"];
        sub_c3_step2[label="Quaståˆ†æ"];
        label="åºåˆ—ç»„è£…åŠç»“æœåˆ†æ";
        sub_c3_step1->sub_c3_step2;
    }

    step1->sub_c1_1;
    step1->sub_c1_2;
    sub_step3->step1;
    sub_c1_3->step2[label="bowtie2 -build"];
    step2->sub_c2_step1;
    sub_c2_step3->sub_c3_step1
}
```

### 1. æ•°æ®å‡†å¤‡

#### 1.1 å‡†å¤‡åŸå§‹åŸºå› ç»„ç»„è£…åºåˆ—

ç”¨ä»¥è¯„ä¼°åç»­çš„åºåˆ—ç»„è£…ç»“æœã€‚

```bash
fnaPath=../../Genomics/data/YJM1342/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna
```

#### 1.2 åˆ©ç”¨art_illuminaæ¨¡æ‹ŸåŒæœ«ç«¯æµ‹åº

æ¨¡æ‹ŸHiSeq2000ï¼Œreadsé•¿åº¦ä¸º100ã€‚è¦†ç›–åº¦å‚æ•°å–19ã€‚
çŸ­æ’å…¥ç‰‡æ®µï¼š

```bash
./art_illumina -ss HS20 -sam -i $fnaPath -p -l 100 -f 19 -m 200 -s 10 -o ../../Genomics/data/YJM1342/Sc_YJM1342_1010_p # çŸ­ç‰‡æ®µ
```

çŸ­æ’å…¥ç‰‡æ®µpairæµ‹åºå›æ˜¾ï¼š

```bash
Paired-end sequencing simulation

Total CPU time used: 50.4272

The random seed for the run: 1570863424

Parameters used during run
Read Length:	100
Genome masking 'N' cutoff frequency: 	1 in 100
Fold Coverage:            10X
Mean Fragment Length:     200
Standard Deviation:       10
Profile Type:             Combined
ID Tag:

Quality Profile(s)
First Read:   HiSeq 2000 Length 100 R1 (built-in profile)
First Read:   HiSeq 2000 Length 100 R2 (built-in profile)

Output files

FASTQ Sequence Files:
the 1st reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_p1.fq
the 2nd reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_p2.fq

ALN Alignment Files:
the 1st reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_p1.aln
the 2nd reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_p2.aln

SAM Alignment File:
../../Genomics/data/YJM1342/Sc_YJM1342_1010_p.sam
```

é•¿æ’å…¥ç‰‡æ®µï¼š

```bash
./art_illumina -ss HS20 -sam -i $fnaPath -mp -l 100 -f 19 -m 2500 -s 50 -o ../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp # é•¿ç‰‡æ®µ
```

é•¿æ’å…¥ç‰‡æ®µmate-pairæµ‹åºå›æ˜¾ï¼š

```bash
Matepair-end sequencing simulation

Total CPU time used: 51.9643

The random seed for the run: 1570863518

Parameters used during run
Read Length:	100
Genome masking 'N' cutoff frequency: 	1 in 100
Fold Coverage:            10X
Mean Fragment Length:     2500
Standard Deviation:       50
Profile Type:             Combined
ID Tag:

Quality Profile(s)
First Read:   HiSeq 2000 Length 100 R1 (built-in profile)
First Read:   HiSeq 2000 Length 100 R2 (built-in profile)

Output files

FASTQ Sequence Files:
the 1st reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp1.fq
the 2nd reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp2.fq

ALN Alignment Files:
the 1st reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp1.aln
the 2nd reads: ../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp2.aln

SAM Alignment File:
../../Genomics/data/YJM1342/Sc_YJM1342_1010_mp.sam

```

#### 1.3 åˆ›å»ºå‚è€ƒåŸºå› ç»„

ç”±äºBowtie2æ¯”å¯¹æ—¶éœ€è¦ç”¨åˆ°çš„å‚è€ƒåŸºå› ç»„ï¼Œç°å¼€å§‹åˆ›å»ºã€‚

```bash
# move files to IBM

bowtie2-build GCA_000977265.3_Sc_YJM1342_v1_genomic.fna Sc_YJM1342_index # æ³¨æ„å·¥ä½œç›®å½•

# å±å¹•å›æ˜¾
Returning block of 2152954
Exited Ebwt loop
fchr[A]: 0
fchr[C]: 3884665
fchr[G]: 6355375
fchr[T]: 8739141
fchr[$]: 12616630
Exiting Ebwt::buildToDisk()
Returning from initFromVector
Wrote 8402213 bytes to primary EBWT file: Sc_YJM1342_index.rev.1.bt2
Wrote 3154164 bytes to secondary EBWT file: Sc_YJM1342_index.rev.2.bt2
Re-opening _in1 and _in2 as input streams
Returning from Ebwt constructor
Headers:
    len: 12616630
    bwtLen: 12616631
    sz: 3154158
    bwtSz: 3154158
    lineRate: 6
    offRate: 4
    offMask: 0xfffffff0
    ftabChars: 10
    eftabLen: 20
    eftabSz: 80
    ftabLen: 1048577
    ftabSz: 4194308
    offsLen: 788540
    offsSz: 3154160
    lineSz: 64
    sideSz: 64
    sideBwtSz: 48
    sideBwtLen: 192
    numSides: 65712
    numLines: 65712
    ebwtTotLen: 4205568
    ebwtTotSz: 4205568
    color: 0
    reverse: 1
Total time for backward call to driver() for mirror index: 00:00:16
```

å¾—åˆ°æ–‡ä»¶å¦‚ä¸‹ï¼š

```bash
(/opt/BioBuilds)-bash-4.2$ ll Sc_YJM1342_index.*
-rw-r--r-- 1 s24 student 8402213 Oct 12 15:24 Sc_YJM1342_index.1.bt2
-rw-r--r-- 1 s24 student 3154164 Oct 12 15:24 Sc_YJM1342_index.2.bt2
-rw-r--r-- 1 s24 student     728 Oct 12 15:23 Sc_YJM1342_index.3.bt2
-rw-r--r-- 1 s24 student 3154158 Oct 12 15:23 Sc_YJM1342_index.4.bt2
-rw-r--r-- 1 s24 student 8402213 Oct 12 15:24 Sc_YJM1342_index.rev.1.bt2
-rw-r--r-- 1 s24 student 3154164 Oct 12 15:24 Sc_YJM1342_index.rev.2.bt2
```

### 2. è´¨æ§åˆ†æ

åˆ©ç”¨fastqcåˆ†åˆ«å¯¹ä¸Šè¿°ä¸¤å¥—é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œè´¨æ§åˆ†æã€‚

```bash
mkdir fastqc_out
fastqc -o ./fastqc_out -f fastq -t 10 Sc_YJM1342_1010_p1.fq  Sc_YJM1342_1010_p2.fq

# å±å¹•å›æ˜¾
Started analysis of Sc_YJM1342_1010_p1.fq
Started analysis of Sc_YJM1342_1010_p2.fq
...
Analysis complete for Sc_YJM1342_1010_p1.fq
Analysis complete for Sc_YJM1342_1010_p2.fq

fastqc -o ./fastqc_out -f fastq -t 10 Sc_YJM1342_1010_mp1.fq  Sc_YJM1342_1010_mp2.fq

# å±å¹•å›æ˜¾
Started analysis of Sc_YJM1342_1010_mp1.fq
Started analysis of Sc_YJM1342_1010_mp2.fq
...
Analysis complete for Sc_YJM1342_1010_mp1.fq
Analysis complete for Sc_YJM1342_1010_mp2.fq

```

### 3. æŸ¥é˜…ä¸¤å¥—æ•°æ®çš„åˆ†æç»“æœå¹¶åˆ†æç»“æœ

#### 3.1 pair1

<table align="center">
  <tr>
    <td>
      <img src="./figs/fastqc/p1/per_base_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p1/per_sequence_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p1/per_base_sequence_content.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p1/per_sequence_gc_content.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342 p1 per_base_quality</td>
    <td>YJM1342 p1 per_sequence_quality</td>
    <td>YJM1342 p1 per_base_sequence_content</td>
    <td>YJM1342 p1 per_sequence_gc_content</td>
  </tr>
  <tr>
    <td><img src="./figs/fastqc/p1/per_base_n_content.png"></img></td>
    <td><img src="./figs/fastqc/p1/sequence_length_distribution.png"></img></td>
    <td><img src="./figs/fastqc/p1/duplication_levels.png"></img></td>
    <td><img src="./figs/fastqc/p1/kmer_profiles.png"></img></td>
  </tr>
  <tr>
    <td>YJM1342 p1 per_base_n_content</td>
    <td>YJM1342 p1 sequence_length_distribution</td>
    <td>YJM1342 p1 duplication_levels</td>
    <td>YJM1342 p1 kmer_profiles</td>
  </tr>
</table>

#### 3.2 pair2

<table align="center">
  <tr>
    <td>
      <img src="./figs/fastqc/p2/per_base_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p2/per_sequence_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p2/per_base_sequence_content.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/p2/per_sequence_gc_content.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342 p2 per_base_quality</td>
    <td>YJM1342 p2 per_sequence_quality</td>
    <td>YJM1342 p2 per_base_sequence_content</td>
    <td>YJM1342 p2 per_sequence_gc_content</td>
  </tr>
  <tr>
    <td><img src="./figs/fastqc/p2/per_base_n_content.png"></img></td>
    <td><img src="./figs/fastqc/p2/sequence_length_distribution.png"></img></td>
    <td><img src="./figs/fastqc/p2/duplication_levels.png"></img></td>
    <td><img src="./figs/fastqc/p2/adapter_content.png"></img></td>
  </tr>
  <tr>
    <td>YJM1342 p2 per_base_n_content</td>
    <td>YJM1342 p2 sequence_length_distribution</td>
    <td>YJM1342 p2 duplication_levels</td>
    <td>YJM1342 p2 adapter_content</td>
  </tr>
</table>

#### 3.3 mate-pair1
<table align="center">
  <tr>
    <td>
      <img src="./figs/fastqc/mp1/per_base_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp1/per_sequence_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp1/per_base_sequence_content.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp1/per_sequence_gc_content.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342 mp1 per_base_quality</td>
    <td>YJM1342 mp1 per_sequence_quality</td>
    <td>YJM1342 mp1 per_base_sequence_content</td>
    <td>YJM1342 mp1 per_sequence_gc_content</td>
  </tr>
  <tr>
    <td><img src="./figs/fastqc/mp1/per_base_n_content.png"></img></td>
    <td><img src="./figs/fastqc/mp1/sequence_length_distribution.png"></img></td>
    <td><img src="./figs/fastqc/mp1/duplication_levels.png"></img></td>
    <td><img src="./figs/fastqc/mp1/adapter_content.png"></img></td>
  </tr>
  <tr>
    <td>YJM1342 mp1 per_base_n_content</td>
    <td>YJM1342 mp1 sequence_length_distribution</td>
    <td>YJM1342 mp1 duplication_levels</td>
    <td>YJM1342 mp1 adapter_content</td>
  </tr>
</table>

#### 3.4 mate-pair2

<table align="center">
  <tr>
    <td>
      <img src="./figs/fastqc/mp2/per_sequence_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp2/per_sequence_quality.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp2/per_base_sequence_content.png"></img>
    </td>
    <td>
      <img src="./figs/fastqc/mp2/per_sequence_gc_content.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342 mp2 per_base_quality</td>
    <td>YJM1342 mp2 per_sequence_quality</td>
    <td>YJM1342 mp2 per_base_sequence_content</td>
    <td>YJM1342 mp2 per_sequence_gc_content</td>
  </tr>
  <tr>
    <td><img src="./figs/fastqc/mp2/per_base_n_content.png"></img></td>
    <td><img src="./figs/fastqc/mp2/sequence_length_distribution.png"></img></td>
    <td><img src="./figs/fastqc/mp2/duplication_levels.png"></img></td>
    <td><img src="./figs/fastqc/mp2/adapter_content.png"></img></td>
  </tr>
  <tr>
    <td>YJM1342 mp2 per_base_n_content</td>
    <td>YJM1342 mp2 sequence_length_distribution</td>
    <td>YJM1342 mp2 duplication_levels</td>
    <td>YJM1342 mp2 adapter_content</td>
  </tr>
</table>

* per_base_quality: å¯ä»¥çœ‹åˆ°åœ¨å¤´å°¾å°¤å…¶æ˜¯å°¾éƒ¨ï¼Œæµ‹åºè´¨é‡è¾ƒå·®ï¼Œä¸­é—´éƒ¨åˆ†ä¸€èˆ¬
* per_sequence_quality: éƒ¨åˆ†è´¨é‡è¾ƒä½ï¼Œä½†æ•°é‡ä¸å¤š
* per_base_sequence_content:å¯ä»¥çœ‹åˆ°ATCGæ¯”ä¾‹å¹³å‡ï¼ŒåŠ å’Œä¸º100%
* per_sequence_gc_contentï¼š å®é™…è®¡ç®—çš„GC contenæ¯”ç†è®ºå€¼ç•¥é«˜
* per_base_n_content: åŸºæœ¬ä¸º0ï¼Œå¯ä»¥çœ‹åˆ°åŸºæœ¬æ²¡æœ‰æœªæµ‹å‡ºç¢±åŸº
* sequence_length_distributionï¼šé›†ä¸­åœ¨100bp
* duplication_levelsï¼šå¯ä»¥çœ‹åˆ°å­˜åœ¨éƒ¨åˆ†å†—ä½™çš„æƒ…å†µ

### 4. æµ‹åºæ•°æ®ä¸å‚è€ƒåŸºå› ç»„çš„æ¯”å¯¹

#### 4.1 åˆ©ç”¨bowtie2ï¼ŒæŠŠä¸Šè¿°ä¸¤å¥—é«˜é€šé‡æµ‹åºæ•°æ®ä¸ä¹‹å‰å‡†å¤‡å¥½çš„åŸºå› ç»„ç´¢å¼•æ–‡ä»¶è¿›è¡Œæ¯”å¯¹ï¼Œä¿ç•™æ¯”å¯¹ç»“æœ

```bash
# download files
# ./fastqc_out/*

bowtie2 -x ./Sc_YJM1342_index -1 Sc_YJM1342_1010_p1.fq,Sc_YJM1342_1010_mp1.fq -2 Sc_YJM1342_1010_p2.fq,Sc_YJM1342_1010_mp2.fq -S ./Sc_YJM1342_1010.sam -p 10 # æ³¨æ„å·¥ä½œç›®å½•

# å±å¹•å›æ˜¾
1260250 reads; of these:
  1260250 (100.00%) were paired; of these:
    630051 (49.99%) aligned concordantly 0 times
    570051 (45.23%) aligned concordantly exactly 1 time
    60148 (4.77%) aligned concordantly >1 times
    ----
    630051 pairs aligned concordantly 0 times; of these:
      554995 (88.09%) aligned discordantly 1 time
    ----
    75056 pairs aligned 0 times concordantly or discordantly; of these:
      150112 mates make up the pairs; of these:
        1158 (0.77%) aligned 0 times
        24974 (16.64%) aligned exactly 1 time
        123980 (82.59%) aligned >1 times
99.95% overall alignment rate

```

å‘ç°åœ¨æœªè®¾ç½®errorå‚æ•°çš„æƒ…å†µä¸‹```overall alignment rate```é100%ï¼Œè¿™æ˜¯ç”±äº```1158 (0.77%) aligned 0 times```ã€‚å…·ä½“åŸå› å¯èƒ½æ˜¯åŸºå› ç»„æ–‡ä»¶ä¸­çš„é‡å¤åºåˆ—é€ æˆé‡å¤readsæ— æ³•å›å¸–çš„é—®é¢˜æˆ–æ˜¯bowtie2è½¯ä»¶ç­‰çš„è¿è¡Œè¾“å‡ºé—®é¢˜ï¼Œè¿˜éœ€æ¢ç©¶ã€‚

#### 4.2 åˆ©ç”¨samtoolså¯¹æ¯”å¯¹ç»“æœå¹¶è¿›è¡Œç®€å•çš„ç»Ÿè®¡åˆ†æ

```bash
samtools view -b  Sc_YJM1342_1010.sam  > Sc_YJM1342_1010.bam #æ ¼å¼è½¬æ¢sam->bam

samtools sort Sc_YJM1342_1010.bam Sc_YJM1342_1010_sorted #æ’åº

# å±å¹•å›æ˜¾
[bam_sort_core] merging from 2 files...

samtools index Sc_YJM1342_1010_sorted.bam #å»ºç«‹ç´¢å¼•

-rw-r--r-- 1 s24 student 857800695 Oct 10 14:07 Sc_YJM1342_1010.sam
-rw-r--r-- 1 s24 student 311328555 Oct 10 14:10 Sc_YJM1342_1010.bam
-rw-r--r-- 1 s24 student 252666394 Oct 10 14:11 Sc_YJM1342_1010_sorted.bam
-rw-r--r-- 1 s24 student     39904 Oct 10 14:12 Sc_YJM1342_1010_sorted.bam.bai

# ç»Ÿè®¡åˆ†æ
samtools stats Sc_YJM1342_1010_sorted.bam > samtools.stat.stats.out

samtools depth Sc_YJM1342_1010_sorted.bam > samtools.stat.depth.out

samtools flagstat Sc_YJM1342_1010_sorted.bam > samtools.stat.flagstat.out

samtools idxstats Sc_YJM1342_1010_sorted.bam > samtools.stat.idxstats.out

# download files
# ./samtools.stat.*

```

#### 4.3 è§£è¯»samtools statsçš„ç»Ÿè®¡ç»“æœ,åˆ©ç”¨plot-bamstatså·¥å…·å¯¹è¾“å‡ºçš„ç»“æœæ–‡æ¡£è¿›è¡Œå¯è§†åŒ–

å¯è§†åŒ–ï¼š

```bash

mkdir plot-bamstats_out #åˆ›å»ºè¾“å‡ºç»“æœå­˜æ”¾ç›®å½•
plot-bamstats -p ./plot-bamstats_out/ ./samtools.stat.stats.out
```

```stat
Reads
total:	2,520,500
filtered:	0	(0.0%)
non-primary:	0
duplicated:	0	(0.0%)
mapped:	2,519,342	(100.0%)
zero MQ:	27	(0.0%)
avg read length:	100
Bases
total:	252,050,000	(100.0%)
mapped:	251,934,200
error rate:	0.96%
```
å¯ä»¥çœ‹åˆ°100%åŒ¹é…ä¸Šï¼Œä½†æ˜¯```error rate```ä¸º0.96%ã€‚

<table align="center">
  <tr>
    <td>
      <img src="./figs/stats/insert-size.png"></img>
    </td>
    <td>
      <img src="./figs/stats/gc-content.png"></img>
    </td>
    <td>
      <img src="./figs/stats/acgt-cycles.png"></img>
    </td>
    <td>
      <img src="./figs/stats/quals.png"></img>
    </td>
  </tr>
  <tr>
    <td>insert-size</td>
    <td>gc-content</td>
    <td>acgt-cycles</td>
    <td>Quality Per Circle 1</td>
  </tr>
  <tr>
    <td><img src="./figs/stats/quals2.png"></img></td>
    <td><img src="./figs/stats/quals3.png"></img></td>
    <td><img src="./figs/stats/quals-hm.png"></img></td>
    <td><img src="./figs/stats/indel-cycles.png"></img></td>
  </tr>
  <tr>
    <td>Quality Per Circle 2</td>
    <td>Quality Per Circle 3</td>
    <td>Quality Per Circle HeatMap</td>
    <td>indel-cycles</td>
  </tr>
  <tr>
    <td><img src="./figs/stats/indel-dist.png"></img></td>
    <td><img src="./figs/stats/gc-depth.png"></img></td>
    <td><img src="./figs/stats/coverage.png"></img></td>
    <td></td>
  </tr>
  <tr>
    <td>indel-dist</td>
    <td>gc-depth</td>
    <td>coverage</td>
    <td></td>
  </tr>
</table>

* gc-contentï¼šé¢‘ç‡åˆ†å¸ƒåˆç†
* acgt-cycles: å¯ä»¥çœ‹åˆ°ATCGæ¯”ä¾‹åŸºæœ¬æ˜¯æ¯”è¾ƒå¹³å‡ï¼ŒåŠ å’Œä¸º100%ï¼Œè¾ƒä¸å¹³ç¨³çš„æ˜¯Gå«é‡
* coverageï¼šå­˜åœ¨éƒ¨åˆ†è¦†ç›–ç‡åœ¨10%ä»¥ä¸‹çš„basesï¼Œå¤§éƒ¨åˆ†é«˜äº15%
* indel-cycles: Deletion(rev)è¾ƒä¸ºæ˜æ˜¾

### 5. åºåˆ—ç»„è£…åŠç»“æœåˆ†æ

é€‰æ‹©SOAPdenovoä½œä¸ºåºåˆ—ç»„è£…è½¯ä»¶ã€‚

#### 5.1 å¯¹ä¸Šè¿°é«˜é€šé‡æµ‹åºæ•°æ®è¿›è¡Œç»„è£…

```bash
# ç¼–å†™é…ç½®æ–‡æ¡£ lib.cfg
max_rd_len=100
[LIB]
avg_ins=200
reverse_seq=0
asm_flags=3
rd_len_cutoff=100
rank=1
pair_num_cutoff=3
map_len=32
q1=Sc_YJM541_1010_p1.fq
q2=Sc_YJM541_1010_p2.fq
[LIB]
avg_ins=2500
reverse_seq=1
asm_flags=3
rank=2
pair_num_cutoff=5
map_len=35
q1=Sc_YJM541_1010_mp1.fq
q2=Sc_YJM541_1010_mp2.fq

nohup SOAPdenovo-63mer all -s lib.cfg -K 30 -o SOAPdenovo_out -p 10 &

(/opt/BioBuilds)-bash-4.2$ ll -o --block-size=M SOAPdenovo_out.*
-rw-r--r-- 1 s24  3M Oct 10 14:38 SOAPdenovo_out.Arc
-rw-r--r-- 1 s24  0M Oct 10 14:39 SOAPdenovo_out.bubbleInScaff
-rw-r--r-- 1 s24 18M Oct 10 14:38 SOAPdenovo_out.contig
-rw-r--r-- 1 s24  1M Oct 10 14:38 SOAPdenovo_out.ContigIndex
-rw-r--r-- 1 s24  1M Oct 10 14:39 SOAPdenovo_out.contigPosInscaff
-rw-r--r-- 1 s24 46M Oct 10 14:37 SOAPdenovo_out.edge.gz
-rw-r--r-- 1 s24  0M Oct 10 14:39 SOAPdenovo_out.gapSeq
-rw-r--r-- 1 s24  1M Oct 10 14:36 SOAPdenovo_out.kmerFreq
-rw-r--r-- 1 s24  3M Oct 10 14:39 SOAPdenovo_out.links
-rw-r--r-- 1 s24  3M Oct 10 14:39 SOAPdenovo_out.newContigIndex
-rw-r--r-- 1 s24  1M Oct 10 14:39 SOAPdenovo_out.peGrads
-rw-r--r-- 1 s24 53M Oct 10 14:37 SOAPdenovo_out.preArc
-rw-r--r-- 1 s24  1M Oct 10 14:37 SOAPdenovo_out.preGraphBasic
-rw-r--r-- 1 s24 13M Oct 10 14:39 SOAPdenovo_out.readInGap.gz
-rw-r--r-- 1 s24 16M Oct 10 14:39 SOAPdenovo_out.readOnContig.gz
-rw-r--r-- 1 s24  3M Oct 10 14:39 SOAPdenovo_out.scaf
-rw-r--r-- 1 s24  1M Oct 10 14:39 SOAPdenovo_out.scaf_gap
-rw-r--r-- 1 s24 12M Oct 10 14:39 SOAPdenovo_out.scafSeq
-rw-r--r-- 1 s24  1M Oct 10 14:39 SOAPdenovo_out.scafStatistics
-rw-r--r-- 1 s24  9M Oct 10 14:38 SOAPdenovo_out.updated.edge
-rw-r--r-- 1 s24 21M Oct 10 14:37 SOAPdenovo_out.vertex
```

#### 5.2 åˆ©ç”¨Quastå°†ç»„è£…ç»“æœä¸­åŒ…å«contigså’Œscaffoldsåºåˆ—çš„æ–‡æ¡£ä¸å‚è€ƒåŸºå› ç»„è¿›è¡Œå¯¹æ¯”

æœ¬åœ°ä¸‹è½½è½¯ä»¶è¿è¡Œï¼š

```bash
input1=/home/student/s24/zeFengZhu/Gen/YJM1342/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna
input2=/home/student/s24/zeFengZhu/Gen/YJM1342/SOAPdenovo_out.contig
input3=/home/student/s24/zeFengZhu/Gen/YJM1342/SOAPdenovo_out.scafSeq
gffPath=/home/student/s24/zeFengZhu/Gen/YJM1342/GCA_000977265.3_Sc_YJM1342_v1_genomic.gff

./quast.py $input1 $input2  -g $gffPath
./quast.py $input1 $input3  -g $gffPath
```

webç«¯è¿è¡Œï¼š(å¿½ç•¥500bpä»¥ä¸‹contigs)

<table align="center">
  <tr>
    <td>
      <img src="./figs/fna_scaf.png"></img>
    </td>
    <td>
      <img src="./figs/fna_contig.png"></img>
    </td>
  </tr>
  <tr>
    <td>fna_scaf QUAST Report</td>
    <td>fna_contig QUAST Report</td>
  </tr>
  <tr>
    <td><img src="./figs/fna_scaf_2.png"></img></td>
    <td><img src="./figs/fna_contig_2.png"></img></td>
  </tr>
  <tr>
    <td>fna_scaf Contig alignment viewer</td>
    <td>fna_contig Contig alignment viewer</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/fna_scaf_3.png"></img>
    </td>
    <td>
      <img src="./figs/fna_contig_3.png"></img>
    </td>
  </tr>
  <tr>
    <td>fna_scaf Contig size viewer</td>
    <td>fna_contig Contig size viewer</td>
  </tr>
</table>

å¯¹å›¾çš„è§£æå‚è§ä¸‹ä¸€æ­¥ã€‚

#### 5.3 æŸ¥çœ‹å¹¶åˆ†ææ¯”å¯¹ç»“æœï¼Œå…³æ³¨å®é™…è¦†ç›–ç‡ç­‰é‡è¦è¯„ä¼°æŒ‡æ ‡

ç»“åˆç†è®ºè®²æˆå†…å®¹å’Œä¸Šä¸€æ­¥ç»Ÿè®¡ç»“æœï¼Œæ¢è®¨å…¶ä¸ç†è®ºè¦†ç›–ç‡é—´çš„å·®å¼‚

* å‚è€ƒåŸºå› ç»„ä¸Scafå’Œcontigçš„å¯¹æ¯”ç»“æœä¸­ï¼Œè¦†ç›–ç‡åˆ†åˆ«ä¸º:86.133%, 74.531%ã€‚
* ä¸ç†è®ºå€¼ç›¸æ¯”ï¼Œä¸Šè¿°å€¼æ˜æ˜¾è¾ƒä½
* å¯èƒ½æ˜¯ç”±äºåŸå§‹æ•°æ®ä¸­çš„è¯¯å·®å¯¼è‡´
* ä¹Ÿå¯èƒ½æ˜¯æ•°æ®ä¸­é‡å¤åºåˆ—è¾ƒå¤šï¼Œå¯¼è‡´æ‹¼æ¥é”™è¯¯æˆ–æŸå¤±ï¼Œéœ€è¿›ä¸€æ­¥ç¡®è®¤

æŸ¥çœ‹è¯¥å¤§æ®µæœªåŒ¹é…åŒºåŸŸï¼š

<table align="center">
  <tr>
    <td>
      <img src="./figs/gap2.png"></img>
    </td>
  </tr>
  <tr>
    <td>gapèµ·å§‹è¾¹ç•Œ</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/gap1.png"></img>
    </td>
  </tr>
  <tr>
    <td>gapç»ˆæ­¢è¾¹ç•Œ</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/gap0.png"></img>
    </td>
  </tr>
  <tr>
    <td>gap Detail</td>
  </tr>
</table>

å¯ä»¥çœ‹åˆ°ï¼Œè¯¥å¤§æ®µæœªåŒ¹é…åŒºåŸŸç”±å¤§é‡ç¼–ç rnaçš„åºåˆ—ç»„æˆï¼Œæå°‘æ•°ä¸ºç¼–ç åŸºå› ï¼Œä¸”å­˜åœ¨å¤§é‡é‡å¤åºåˆ—ã€‚é€ æˆäº†ç»„è£…çš„éƒ¨åˆ†åŒºåŸŸå¤±è´¥ã€‚

## å®éªŒé¡¹ç›®3: åŸºå› ç»„æ³¨é‡Šä¹‹åŒæºæœç´¢

### å®éªŒç›®çš„

1. åŠ æ·±åŒæºæœç´¢ç›¸å…³ç†è®ºçŸ¥è¯†çš„ç†è§£
2. æŒæ¡å…¨åŸºå› ç»„çš„åŒæºåŸºå› é¢„æµ‹æ–¹æ³•
3. æŒæ¡ç›¸å…³æ•°æ®åº“å’Œè½¯ä»¶çš„ä½¿ç”¨

### å®éªŒæµç¨‹

```viz
digraph flowchart_3 {
    # rankdir=LR;
    fontname="Courier New";
    size="6,4"; ratio = fill;
    node [style="filled,setlinewidth(3)", color="#8383cc", fontname="Courier New", shape="Mrecord",fixedsize=true,width=2.5,fillcolor="#d9e7ee"];
    edge [color="0.635 0.707 0.707", fontname="Courier New"];
    label="åŸºå› ç»„æ³¨é‡Šä¹‹åŒæºæœç´¢";
    # step1[label="æ•°æ®å‡†å¤‡åŠé¢„å¤„ç†"];
    step2[label="åˆ›å»ºæœ¬åœ°BLAST DB"];
    step3[label="å…¨åŸºå› ç»„åŒæºåŸºå› æœç´¢"];
    step4[label="åŒæºæœç´¢ç»“æœè¯„ä¼°"];

    subgraph cluster_1{
        style=filled;
        color=lightgrey;
        node [color=white];
        label="æ•°æ®å‡†å¤‡åŠé¢„å¤„ç†"
        sub_c1_1[label="åŸºå› ç»„åºåˆ—"];
        sub_c1_2[label="å·²çŸ¥è›‹ç™½åºåˆ—"];
    }

    sub_c2_0[label="åŸå§‹GFFæ–‡æ¡£", color=white];
    sub_c1_1->step2[label="makeblastdb"];
    step2->step3;
    sub_c1_2->step3[label="tblastn"];
    "code1" [ style = "filled" penwidth = 1 fillcolor = "white" fontname = "Courier New" shape = "Mrecord" label =<<table border="0" cellborder="0" cellpadding="3" bgcolor="white"><tr><td bgcolor="black" align="center" colspan="2"><font color="white">blast92gff3.pl</font></td></tr><tr><td align="left" port="r3">perl code</td></tr></table>> ];
    "code2" [ style = "filled" penwidth = 1 fillcolor = "white" fontname = "Courier New" shape = "Mrecord" label =<<table border="0" cellborder="0" cellpadding="3" bgcolor="white"><tr><td bgcolor="black" align="center" colspan="2"><font color="white">blast2gff.py</font></td></tr><tr><td align="left" port="r3">python code</td></tr></table>> ];
    step3->code1;
    step3->code2;
    code1->step4;
    code2->step4;
    sub_c2_0->step4[label="gffcompare"];
}
```

### 1. æ•°æ®å‡†å¤‡åŠé¢„å¤„ç†

#### 1.1 åŸºå› ç»„åºåˆ—

> å®éªŒé¡¹ç›® 1 ä¸­ä¸‹è½½æˆ–ç»„è£…æˆ–ç”±ä»»è¯¾æ•™å¸ˆæä¾›çš„åŸºå› ç»„åºåˆ—ï¼ˆFASTA æ ¼å¼ï¼‰. ä»¥åŠç›¸åº”çš„ GFF æ ¼å¼æ³¨é‡Šæ–‡ä»¶

```bash
workDir="/home/student/s24/zeFengZhu/Gen/lab3"
fastaFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna"
gffFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.gff"
```

#### 1.2 å·²çŸ¥è›‹ç™½åºåˆ—

> æ ¹æ®åŸºå› ç»„åºåˆ—çš„ç‰©ç§æ¥æºï¼Œä» UniProt æ•°æ®åº“æœç´¢. ä¸‹è½½è¿‘ç¼˜ç‰©ç§æ‰€æœ‰å·²çŸ¥è›‹ç™½åºåˆ—(reviewed), ä¿å­˜åºåˆ—æ¡æ•°ä»¥åŠ FASTA æ ¼å¼åºåˆ—ï¼Œç”¨äºå…¨åŸºå› ç»„çš„åŒæºæœç´¢

è¿›å…¥UniProtè¿›è¡Œæœç´¢ï¼š

```bash
# æœç´¢å†…å®¹
taxonomy:fungi NOT "saccharomyces cerevisiae" AND reviewed:yes
```

ä¸‹è½½fastaæ ¼å¼æ–‡ä»¶ï¼š

```bash
# å¾—åˆ°æ–‡ä»¶
uniprot-taxonomy_fungi+NOT+_saccharomyces+cerevisiae_+AND+reviewed_yes.fasta
# é‡å‘½åä¸º
protein.fasta
# è®¾ç½®è·¯å¾„
unpFastaFile="$workDir/protein.fasta"
```

### 2 åˆ›å»ºæœ¬åœ° BLAST æ•°æ®åº“

> è¿œç¨‹ç™»å½• IBM å°å‹æœºï¼Œä½¿ç”¨ makeblastdb ç¨‹åºï¼Œå¯¹ä¸Šè¿° FASTA æ ¼å¼çš„åŸºå› ç»„åºåˆ—è¿›è¡Œå¤„ç†ï¼Œå»ºç«‹æœ¬åœ° BLAST æ•°æ®åº“ã€‚

```bash
makeblastdb -in $fastaFile -input_type fasta -title Sc_gDNA -dbtype nucl -out Sc_gDNA
```

è¾“å‡ºå¦‚ä¸‹ï¼š

```bash
Building a new DB, current time: 10/24/2019 14:22:05
New DB name:   /home/student/s24/zeFengZhu/Gen/lab3/Sc_gDNA
New DB title:  Sc_gDNA
Sequence type: Nucleotide
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 17 sequences in 0.468456 seconds.
```

### 3å…¨åŸºå› ç»„çš„åŒæºåŸºå› æœç´¢

#### 3.1 ä½¿ç”¨ tblastn ç¨‹åº, æŠŠå·²çŸ¥è›‹ç™½è´¨åºåˆ—å’Œä¸Šè¿°å»ºç«‹çš„æœ¬åœ° BLAST æ•°æ®åº“è¿›è¡Œæ¯”å¯¹

> æ³¨æ„å‚æ•°è®¾ç½® (å¦‚ e-value è®¾ä¸º 0.00001ï¼Œå»ºè®®è¾“å‡ºæ ¼å¼ 6 å’Œ 7ï¼Œæ ¼å¼ 6 ä¸º tabular æ ¼å¼,7 ç›¸å¯¹äº 6 æ›´ä¸ºè¯¦ç»†ï¼Œå¯ä»¥æŸ¥çœ‹ 6 ä¸­ç¼ºå¤±çš„è¡¨æ ¼æ ‡é¢˜ä¿¡æ¯)



ç¼–å†™shellè„šæœ¬ï¼š
æ³¨æ„åœ¨è¿œç¨‹è¿æ¥æœåŠ¡å™¨ç¼–è¾‘shellè„šæœ¬æ—¶ï¼Œè¦æå‰æ”¹å¥½æ¢è¡Œç¬¦(``:set ff=unix``)ï¼Œåˆ‡æ¢åˆ°UNIXæ ¼å¼,å¦åˆ™shellè„šæœ¬æ— æ³•æ­£å¸¸è¿è¡Œ

```bash
# å…¨åŸºå› ç»„çš„åŒæºåŸºå› æœç´¢
tblastn -query $unpFastaFile -db SC_gDNA -out ./blastx_results.outfmt6 -evalue 1e-5 -outfmt 6 -max_target_seqs 1 -num_threads 10;
tblastn -query $unpFastaFile -db SC_gDNA -out ./blastx_results.outfmt7 -evalue 1e-5 -outfmt 7 -max_target_seqs 1 -num_threads 10;
```

å¾—åˆ°æ–‡ä»¶

```bash
-rw-r--r-- 1 s24 student  2127292 Oct 24 18:15 blastx_results.outfmt6
-rw-r--r-- 1 s24 student  9984620 Oct 24 15:37 blastx_results.outfmt7
```

#### 3.2 ä½¿ç”¨ blast92gff3.pl å’Œ blast2gff.py ç¨‹åºï¼Œåˆ†åˆ«æŠŠç»“æœè½¬æˆ GFF3 æ ¼å¼

ç¨‹åºæ¥æº:

```bash
http://bioinformatics.suda.edu.cn/teaching/2018-2019-2/genomics/tools/blast2gff.py
http://bioinformatics.suda.edu.cn/teaching/2018-2019-2/genomics/tools/blast92gff3.pl
```

è¿è¡Œç¨‹åº:

```bash
perl blast92gff3.pl blastx_results.outfmt6 > Sc_perl.gff3
# å›æ˜¾:
# Summary of HSPs saved
# ALL saved = 26119
# other saved = 26119

python2 blast2gff.py -b blastx_results.outfmt6 > Sc_python.gff3
# å›æ˜¾:
Starting BLAST parsing on blastx_results.outfmt6 Thu Oct 25 08:19:43 2019
Parsed 26780 lines Thu Oct 25 08:19:44 2019
Found 13001 forward and 13779 reverse hits Thu Oct 25 08:19:44 2019
Wrote 26780 matches Thu Oct 25 08:19:44 2019
```

#### 3.3 æ¯”è¾ƒä¸¤ä¸ªç¨‹åºè½¬æ¢ç»“æœçš„å¼‚åŒä¹‹å¤„

|æ¡ç›®|Sc_python|Sc_perl|äº¤é›†|
-|-|-|-
|è®°å½•æ•°/è¡Œæ•°|26780|47027|-|
|èµ·å§‹ä½ç‚¹é›†å†…æ•°ç›®|13688|12794|12794|
|ç»ˆæ­¢ä½ç‚¹é›†å†…æ•°ç›®|13864|12908|12908|

* ``Sc_python.gff3``å…±æœ‰26780æ¡è®°å½•
* ``Sc_perl.gff3``å…±æœ‰47027æ¡è®°å½•,æ¯”å‰è€…å¤šè¿‘ä¸€å€ï¼ŒåŸå› åœ¨äºæ¯ä¸ªmatchåŸºæœ¬éƒ½æœ‰HSPä¸ä¹‹å¯¹åº”,æœ‰å¤šå¤–æ˜¾å­

* äºŒè€…Matchä¸Šçš„èµ·å§‹ä½ç‚¹é›†éƒ½æ¯”å…¶ç»ˆæ­¢ä½ç‚¹é›†å¤š
* ``Sc_python.gff3``å†…çš„ä½ç‚¹é›†åŒ…å«äº†``Sc_perl.gff3``çš„ä½ç‚¹é›†, ä¹Ÿå°±æ˜¯è¯´``Sc_python.gff3``æœªç»è¿‡åˆ†å€¼çš„ç­›é€‰
#####


#### 3.4 æ’é™¤ blast æ¯”å¯¹ç»“æœä¸­çš„å†—ä½™é¡¹

1. ä¸åŒç‰©ç§çš„åŒæºè›‹ç™½åœ¨åŸºå› ç»„ä¸Šçš„åŒ¹é…ä½ç½®å­˜åœ¨çš„é‡å é—®é¢˜ $\rightarrow$ é€šè¿‡å¯¹ç‰©ç§è¿›è¡Œåˆ†ç»„è§£å†³
2. åŒä¸€è›‹ç™½å®¶æ—çš„ä¸åŒæˆå‘˜åœ¨åŸºå› ç»„ä¸Šçš„åŒ¹é…ä½ç½®å­˜åœ¨çš„é‡å é—®é¢˜ $\rightarrow$ é€šè¿‡å¯¹è›‹ç™½å®¶æ—\&æˆå‘˜ è¿›è¡Œåˆ†ç»„è§£å†³
3. åŒä¸€ä¸ªè›‹ç™½åœ¨åŸºå› ç»„ä¸Šçš„ä¸åŒä½ç½®çš„é«˜ç›¸ä¼¼åŒºåŸŸé—®é¢˜ $\rightarrow$ é€šè¿‡è´ªå©ªç®—æ³•é€‰å–è¦†ç›–ç‡æœ€é«˜å¹¶å½¼æ­¤äº’ä¸äº§ç”Ÿè¶…è¿‡é˜ˆå€¼çš„äº¤é›†çš„åŒ¹é…ç‰‡æ®µ

```py
example = "Parent=P61872|LIP_RHIOR_G1;Target=sp:P61872|LIP_RHIOR 187 348;align=198" # of perl
example2 = "ID=spP87050CDR2_SCHPO.exon4;Target=spP87050CDR2_SCHPO 44 209" # of python
pattern = "Target=[A-z0-9:]+\|([A-z0-9_]+)\s"
pattern2 = "Target=sp[A-z0-9]{6}([A-z0-9_]+)\s"
```

### 4 åŒæºæœç´¢ç»“æœçš„è¯„ä¼°
> ä½¿ç”¨ gffcompare å·¥å…·æŠŠç¬¬ 4 æ­¥ç»“æœä¸ 1.1 æ­¥åŸå§‹ GFF æ ¼å¼æ•°æ®è¿›è¡Œæ¯”è¾ƒï¼ŒæŸ¥çœ‹ç»“æœï¼Œå¹¶åˆ†æå®ƒä»¬ä¹‹é—´çš„å¼‚åŒ

ä»[GitHubgffcompare](https://github.com/gpertea/gffcompare "Link")ä¸‹è½½``gffcompare``

è¿›è¡Œæ¯”è¾ƒ

```python
#æŠŠ Sc_python.gff3 æ–‡æ¡£ä¸­çš„ match_part æ›¿æ¢æˆ exon
>>> with open("Sc_python.gff3","rt") as inFile:
...     with open("Sc_python_modified.gff3","wt") as outFile:
...             for line in inFile:
...                     outFile.write(line.replace("match_part","exon"))
#æŠŠ Sc_perl.gff3 æ–‡æ¡£ä¸­çš„ match æ›¿æ¢æˆ geneï¼ŒHSP æ›¿æ¢æˆ exon
>>> with open("Sc_perl.gff3","rt") as inFile:
...     with open("Sc_perl_modified.gff3","wt") as outFile:
...             for line in inFile:
...                     outFile.write(line.replace("match","gene").replace("HSP", "exon"))
```

```bash
gffcompare -V -R -r $gffFile ./Sc_perl_modified.gff3 -o ./Sc_perl
gffcompare -V -R -r $gffFile ./Sc_python_modified.gff3 -o ./Sc_python
```

è¾“å‡º:

```bash
-rw-r--r-- 1 s24 student 8416134 Oct 31 09:26 Sc_perl.annotated.gtf
-rw-r--r-- 1 s24 student  681247 Oct 31 09:26 Sc_perl.loci
-rw-r--r-- 1 s24 student  664799 Oct 31 09:26 Sc_perl.Sc_perl_modified.gff3.refmap
-rw-r--r-- 1 s24 student 2658224 Oct 31 09:26 Sc_perl.Sc_perl_modified.gff3.tmap
-rw-r--r-- 1 s24 student    1326 Oct 31 09:26 Sc_perl.stats
-rw-r--r-- 1 s24 student 2750949 Oct 31 09:26 Sc_perl.tracking
-rw-r--r-- 1 s24 student 9279173 Oct 31 09:26 Sc_python.annotated.gtf
-rw-r--r-- 1 s24 student  875844 Oct 31 09:26 Sc_python.loci
-rw-r--r-- 1 s24 student  956712 Oct 31 09:26 Sc_python.Sc_python_modified.gff3.refmap
-rw-r--r-- 1 s24 student 3451053 Oct 31 09:26 Sc_python.Sc_python_modified.gff3.tmap
-rw-r--r-- 1 s24 student    1283 Oct 31 09:26 Sc_python.stats
-rw-r--r-- 1 s24 student 3452481 Oct 31 09:26 Sc_python.tracking
```

|Sc_perl| Sensitivity | Precision  |Sc_python|Sensitivity|Precision|
-|-|-|-|-|-
Base level|	76.7|	96.3|-|76.6|96.3
Exon level|	46.5|	30.5|-|46.3|28.1
Intro level|	0.5|	0.1|-|0.0     |     nan
Intro chain level|	0.6|	0.1|-|0.0     |     nan
Transcript level|	60.2|	9.6|-|60.1     |     8.2
Locus level|	60.3|	56.6|-|60.2     |    53.2
-|-|-|-|-|-
Matching intron chains|1||-|0
Matching transcripts|2179||-|2185
Matching loci|2179||-|2185
-|-|-|-|-|-
Missed exons|	159/3803|(4.2%)|-|161/3824|(4.2%)
Novel exons|	587/17753|(3.3%)|-|628/19382|(3.2%)
Missed introns|	160/184|(87.0%)|-|186/186|(100.0%)
Novel introns|	1096/1151|(95.2%)|-||
Missed loci|	0/3611|(0.0%)|-|0/3630|(0.0%)
Novel loci|	223/3852|(5.8%)|-|234/4109|(5.7%)

* Base Levelï¼šåœ¨ç›¸åŒåæ ‡ä¸ŠæŠ¥å‘Šçš„å¤–æ˜¾å­ç¢±åŸºçš„æ•°ç›®æƒ…å†µ
  * Sensitivityï¼šéƒ½çº¦ä¸º78%ï¼Œè¯´æ˜BLASTç»“æœåœ¨è¯¥æ°´å¹³ä¸Šç»“æœæ‰¾åˆ°äº†ä¸å°‘åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£ä¸­çš„å†…å®¹ï¼Œä½†æœ‰ä¸€éƒ¨åˆ†ç¢±åŸºæ ¹æœ¬æ²¡æœ‰è¢«ä»»ä½•é¢„æµ‹çš„è½¬å½•æœ¬ï¼ˆtransfragsï¼‰å¤–æ˜¾å­æ‰€è¦†ç›–
  * Precisionï¼šé«˜è¾¾96.3%ï¼Œè¯´æ˜BLASTç»“æœåœ¨è¯¥æ°´å¹³ä¸Šä»…æœ‰ä¸€å°éƒ¨åˆ†ï¼ˆ3.7%ï¼‰ç¢±åŸºè¢«é¢„æµ‹çš„è½¬å½•æœ¬å¤–æ˜¾å­è¦†ç›–ä½†æœªè¢«ä»»ä½•å‚è€ƒè½¬å½•æœ¬å¤–æ˜¾å­è¦†ç›–
* Exon levelï¼šä¸¤æ–‡ä»¶åŸºå› ç»„ä¸Šçš„å¤–æ˜¾å­é—´éš”äº¤é›†æƒ…å†µ
  * å¯ä»¥çœ‹åˆ°é¢„æµ‹åŸºå› ç»“æœçš„å¤–æ˜¾å­ä¸åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£çš„å¤–æ˜¾å­è¾¹ç•Œæœ‰ä¸€å®šå°å·®å¼‚
* Intron levelï¼šå†…å«å­é—´éš”
  * é¢„æµ‹åŸºå› çš„å†…å«å­è¾¹ç•Œæœ‰å¤§é‡ä¸åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£å­˜åœ¨å·®å¼‚ï¼Œä¸”é”™è¯¯é¢„æµ‹äº†æ›´å¤šå†…å«å­ï¼ŒPrecision. Sensitivityæ•°å€¼è¿‡ä½
* Transcript levelï¼šé¢„æµ‹è½¬å½•æœ¬ä¸å‚è€ƒè½¬å½•æœ¬é—´çš„åŒ¹é…æƒ…å†µ
  * è½¬å½•æ°´å¹³åŒ¹é…å­˜åœ¨ä¸å°‘"è¯¯å·®"
* Locus levelï¼šè§‚å¯Ÿåˆ°çš„åŸºå› åº§ï¼ˆå¤–æ˜¾å­é‡å çš„è½¬å½•ç‰©ç°‡ï¼‰ä¸æ„å»ºçš„å‚è€ƒåŸºå› åº§çš„ç›¸ä¼¼åŒ¹é…æƒ…å†µ
  * åŸºå› åº§ä½ç½®åŒ¹é…å­˜åœ¨ä¸å°‘"è¯¯å·®"

1. Perlç¨‹åºè¿è¡Œç»“æœæ™®éæ¯”Pythonè¿è¡Œç»“æœå¥½çš„æ ¹æœ¬åŸå› åœ¨äºå…¶å¯¹åˆ†å€¼è¿›è¡Œäº†ç­›é€‰ï¼Œè€Œåè€…æ²¡æœ‰
2. å¯¹æ¯”ç»“æœä¸­å­˜åœ¨æ–°å¢å¤–æ˜¾å­. å†…å«å­ï¼Œä¸ºå‡é˜³æ€§ã€‚åŸå› å¯èƒ½æ˜¯åœ¨ BLASTè¿‡ç¨‹ä¸­æŸäº›è¿‘æºç‰©ç§çš„è›‹ç™½è´¨å¯¹æ¯”æˆåŠŸã€‚
2. ä¸¢å¤±å†…å«å­å¤–æ˜¾å­çš„åŸå› å¯èƒ½æ˜¯ BLASTè¿‡ç¨‹ä¸­æ²¡æœ‰æ¯”å¯¹ä¸Šæˆ–è€…æ˜¯ç”±äº BLASTå¯å‘å¼ç®—æ³•é—®é¢˜å¯¼è‡´çš„é—æ¼ã€‚
3. ç”±äºå¯¹æ¯”ææ–™ä¸­æ–°å¢äº†è¿‘æºç‰©ç§çš„è›‹ç™½è´¨åºåˆ—ï¼Œå¯¹æ¯”ç»“æœçš„éƒ¨åˆ†æ•°å€¼ç†è®ºä¸Šæ˜¯åé«˜çš„ã€‚

## å®éªŒé¡¹ç›®4: åŸºå› ç»„æ³¨é‡Šä¹‹ä»å¤´é¢„æµ‹ä¸ç»“æ„å»ºæ¨¡

### å®éªŒç›®çš„

1. åŠ æ·±ä»å¤´åŸºå› é¢„æµ‹å’ŒåŸºå› ç»“æ„å»ºæ¨¡ç›¸å…³ç†è®ºçŸ¥è¯†çš„ç†è§£
2. ç†Ÿæ‚‰å’ŒæŒæ¡åŸºäºä»å¤´é¢„æµ‹æ–¹æ³•çš„åŸºå› é¢„æµ‹è½¯ä»¶çš„ä½¿ç”¨

### å®éªŒæµç¨‹

```viz
digraph flowchart_4 {
    # rankdir=LR;
    fontname="Courier New";
    size="6,5"; ratio = fill;
    node [style="filled,setlinewidth(3)", color="#8383cc", fontname="Courier New", shape="Mrecord",fixedsize=true,width=2.5,fillcolor="#d9e7ee"];
    edge [color="0.635 0.707 0.707", fontname="Courier New"];
    label="åŸºå› ç»„æ³¨é‡Šä¹‹ä»å¤´é¢„æµ‹ä¸ç»“æ„å»ºæ¨¡";
    # step1[label="æ•°æ®å‡†å¤‡"];
    step2[label="å…¨åŸºå› ç»„ä»å¤´åŸºå› é¢„æµ‹"];
    # step3[label="ä»å¤´åŸºå› é¢„æµ‹ç»“æœçš„é‰´åˆ«"];
    step4[label="ä»å¤´é¢„æµ‹ç»“æœçš„è¯„ä¼°"];

    subgraph cluster_1{
        style=filled;
        color=lightgrey;
        node [color=white];
        label="æ•°æ®å‡†å¤‡"
        sub_c1_1[label="åŸºå› ç»„åºåˆ—"];
        sub_c1_2[label="å·²çŸ¥è›‹ç™½åºåˆ—"];
        sub_c1_3[label="åŸå§‹GFFæ–‡æ¡£"];
    }

    subgraph cluster_2{
        style=filled;
        color=lightgrey;
        node [color=white];
        label="ä»å¤´åŸºå› é¢„æµ‹ç»“æœçš„é‰´åˆ«"
        sub_c2_1[label="åˆ›å»ºæœ¬åœ°BLAST DB"];
        sub_c2_2[label="æå–è›‹ç™½åºåˆ—"];
        sub_c2_3[label="é‰´åˆ«é¢„æµ‹å‡ºæ¥çš„åŸºå› "];
        sub_c2_4[label="é¢„æµ‹ç»“æœåˆå¹¶"];
    }

    sub_c1_1->step2[label="Augustus"];
    sub_c1_2->sub_c2_1[label="makeblastdb"];
    step2->sub_c2_2[label="GFF file"];
    sub_c2_2->sub_c2_3;
    sub_c2_1->sub_c2_3;
    sub_c2_3->sub_c2_4;
    step2->sub_c2_4[label="GFF file"];
    sub_c2_4->step4[label="gffcompare"];
    # sub_c1_3->step4;
}
```

### 1. åŸºå› ç»„æ•°æ®å‡†å¤‡

> å®éªŒé¡¹ç›® 1 ä¸­ä¸‹è½½æˆ–ç»„è£…æˆ–ç”±ä»»è¯¾æ•™å¸ˆæä¾›çš„åŸºå› ç»„åºåˆ—ï¼ˆFASTA æ ¼å¼ï¼‰. ä»¥åŠç›¸åº”çš„ GFF æ ¼å¼æ³¨é‡Šæ–‡ä»¶

```bash
workDir="/home/student/s24/zeFengZhu/Gen/lab4"
fastaFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna"
gffFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.gff"
```

### 2. ä»å¤´åŸºå› é¢„æµ‹è½¯ä»¶çš„å®‰è£…ä¸æµ‹è¯•

> ä»ç½‘ä¸Šæœç´¢. ä¸‹è½½å¹¶å®‰è£…åŸºå› é¢„æµ‹ç›¸å…³è½¯ä»¶ï¼Œå¦‚ Augustus. GeneMarkES/ET ç­‰

å¯ç”¨åœ¨Anacondaç¯å¢ƒä¸‹åˆ©ç”¨Biocondaå®‰è£…augustus,ä½†ä»…é’ˆå¯¹Linux-64/Mac OSX-64ç³»ç»Ÿè€Œä¸æ”¯æŒWindowsç³»ç»Ÿã€‚

> Bioconda is a distribution of bioinformatics software realized as a channel for the versatile Conda package manager. (https://anaconda.org/bioconda/augustus)

```bash
conda install -c bioconda augustus
```

### 3. å…¨åŸºå› ç»„çš„ä»å¤´åŸºå› é¢„æµ‹

#### 3.1 ä»»é€‰ä¸€ä¸ªèƒ½å¤Ÿè¿›è¡Œä»å¤´é¢„æµ‹åŸºå› çš„è½¯ä»¶ï¼Œå¦‚ Augustus. GeneMarkES/ET ç­‰

é€‰æ‹©Augustusè¿›è¡Œå®éªŒã€‚

```bash
augustus --gff3=on --outfile=Sc_augustus_out.gff3 --species=saccharomyces_cerevisiae_S288C $fastaFile
augustus --species=saccharomyces_cerevisiae_S288C --UTR=off --strand=both --sample=100 --keep_viterbi=true --alternatives-from-sampling=false --genemodel=partial /data/www/augpred/webdata/pred9dTmEkZ9/genome.fa --codingseq=on --exonnames=on

```

> å¯¹å‚æ•°å­˜ç–‘

#### 3.2 ä½¿ç”¨è¯¥è½¯ä»¶å¯¹ç¬¬ 1 æ­¥å‡†å¤‡çš„åŸºå› ç»„åºåˆ—è¿›è¡ŒåŸºå› é¢„æµ‹åˆ†æï¼Œ

ä¿å­˜ GFF æ ¼å¼çš„é¢„æµ‹ç»“æœï¼Œä»¥åŠç›¸åº”çš„å¤šè‚½æˆ– CDS åºåˆ—ï¼ˆFASTA æ ¼å¼ï¼‰
å¾—åˆ°æ–‡ä»¶ï¼š

```bash
augustus.aa
augustus.cdsexons
augustus.codingseq
augustus.gbrowse
augustus.gff
augustus.gtf
```

### 4 ä»å¤´åŸºå› é¢„æµ‹ç»“æœçš„é‰´åˆ«

> æ³¨ï¼š 4.1æ­¥éª¤åœ¨å®éªŒä¸‰ä¸­å·²ç»å®Œæˆï¼Œæœ¬æ¬¡å®éªŒç›´æ¥é‡‡ç”¨å®éªŒä¸‰æ–‡ä»¶ï¼Œä¸‹è¿°å†ä¸€æ¬¡è®°å½•ç›¸å…³æ­¥éª¤ã€‚

#### 4.1 å·²çŸ¥è›‹ç™½åºåˆ—

æ ¹æ®åŸºå› ç»„åºåˆ—çš„ç‰©ç§æ¥æºï¼Œä» UniProt æ•°æ®åº“æœç´¢. ä¸‹è½½è¿‘ç¼˜ç‰©ç§æ‰€æœ‰å·²çŸ¥è›‹ç™½åºåˆ—ï¼ˆreviewedï¼‰

è¿›å…¥UniProtè¿›è¡Œæœç´¢ï¼š

```bash
# æœç´¢å†…å®¹
taxonomy:fungi NOT "saccharomyces cerevisiae" AND reviewed:yes
```

ä¸‹è½½fastaæ ¼å¼æ–‡ä»¶ï¼š

```bash
# å¾—åˆ°æ–‡ä»¶
uniprot-taxonomy_fungi+NOT+_saccharomyces+cerevisiae_+AND+reviewed_yes.fasta
# é‡å‘½åä¸º
protein.fasta
# è®¾ç½®è·¯å¾„
unpFastaFile="$workDir/protein.fasta"
```

#### 4.2 åˆ›å»ºæœ¬åœ° BLAST æ•°æ®åº“

ä½¿ç”¨ makeblastdb ç¨‹åºï¼Œå¯¹ä¸Šè¿° FASTA æ ¼å¼çš„è›‹ç™½è´¨åºåˆ—è¿›è¡Œå¤„ç†ï¼Œå»ºç«‹æœ¬åœ° BLAST æ•°æ®åº“

```bash
makeblastdb -in $unpFastaFile -input_type fasta -title uniprot_protein -dbtype prot -out uniprot_protein
```

è¾“å‡ºå¦‚ä¸‹ï¼š

```bash
Building a new DB, current time: 11/14/2019 17:06:57
New DB name:   /home/student/s24/zeFengZhu/Gen/lab4/uniprot_protein
New DB title:  uniprot_protein
Sequence type: Protein
Keep Linkouts: T
Keep MBits: T
Maximum file size: 1000000000B
Adding sequences from FASTA; added 24905 sequences in 2.86177 seconds.
```

#### 4.3 ä»GFFæ–‡æ¡£ä¸­æå–FASTAåºåˆ—

##### GFFä¸­åºåˆ—æ ¼å¼èŒƒä¾‹

```bash
# start gene g1
...
# protein sequence = [MVKLTSIAAGQITSSITSSRPIITPFYPSNGTSVISSSVISSSVISSSVTSSL...
# SIFSESS...
...
# TTEITKQTTETTKQTTETTKQTTVVTIFSCESDVCSKTASPAIVSTSTATINDVTTEYTTWCPISTTESRQQT...]
# end gene g1
```

å¯ä»¥çœ‹åˆ°ï¼ŒFASTAåºåˆ—è®°å½•çš„æ¨¡å¼å¯ä»¥æ€»ç»“ä¸º:

```python
startwith = "# protein sequence = \[([A-z]+)" # # coding sequence = \[([a-z]+)
content = "([A-z]+)"
endwith = "# ([A-z]+)]"
endkey = "end gene ([A-z0-9]+)"
```

##### æå–åºåˆ—å‡½æ•°

```python
def ExtractSeqFromGFF3(text, startwith=r"# protein sequence = \[([A-z]+)", content="([A-z]+)", endwith="([A-z]+)]", endKey="end gene ([A-z0-9]+)"):
    assert isinstance(text, (Iterable, Iterator)), "Invalid Object"

    startwith, content, endwith, endKey = (re.compile(i) for i in (startwith, content, endwith, endKey))
    flag, endToken, seq = 0, 0, ""

    for line in text:
        startToken = startwith.search(line)
        if startToken is not None:
            flag = 1
            seq += startToken.group(1)
        if flag:
            endToken = endwith.search(line)
            if endToken is not None:
                flag = 0
            if startToken is None:
                seq += content.search(line).group(1)
        elif endToken is not None:
            key = endKey.search(line)
            if key is not None:
                yield key.group(1), seq
                endToken, seq = 0, ""


def toFASTA(name, seq):
    return ">{name}\n{seq}\n".format(name=name, seq=seq)


def script(inPath, outPath, mode):
    with open(inPath, "rt") as inFile:
        with open(outPath, "wt") as outFile:
            if mode == "gene":
                g = ExtractSeqFromGFF3(inFile, startwith=r"# coding sequence = \[([a-z]+)")
            else:
                g = ExtractSeqFromGFF3(inFile)
            for name, seq in g:
                outFile.write(toFASTA(name, seq[:-1]))


if __name__ == "__main__":
    script("Sc_augustus_out.gff3", "augustus_gene.fasta", "gene")
    script("Sc_augustus_out.gff3", "augustus_protein.fasta", "protein")
```

æå–å‡ºé¢„æµ‹åŸºå› åºåˆ—æ–‡ä»¶ï¼š```augustus_gene.fasta```; è›‹ç™½åºåˆ—æ–‡ä»¶ï¼š```augustus_protein.fasta```

#### 4.4 ä½¿ç”¨åˆé€‚çš„ blast ç¨‹åºå¯¹è¯¥é¢„æµ‹åŸºå› ä¸å·²çŸ¥è›‹ç™½åºåˆ—è¿›è¡Œæ¯”å¯¹,ä»¥æ­¤æ¥é‰´åˆ«ä»å¤´é¢„æµ‹å‡ºæ¥çš„åŸºå› 

> åªä¿ç•™æ‰“åˆ†æœ€é«˜çš„ä¸€æ¡ç»“æœï¼Œç”±```max_target_seqs```æŒ‡å®š

```bash
nohup blastx -query ./augustus_gene.fasta -db uniprot_protein -out ./Sc_blastx_gene_results.outfmt6 -evalue 1e-5 -outfmt 6 -max_target_seqs 1 -num_threads 10 > nohup_blastx_gene.out &
# æˆ–
nohup blastp -query ./augustus_protein.fasta -db uniprot_protein -out ./Sc_blastp_gene_results.outfmt6 -evalue 1e-5 -outfmt 6 -max_target_seqs 1 -num_threads 10 > nohup_blastp_gene.out &
```


#### 4.5 æŠŠ 4.4 ç»“æœåˆå¹¶åˆ° 3.2 è·å¾—çš„ GFF æ ¼å¼ç»“æœä¸­

ä½¿ç”¨ç›¸ä¼¼æ€§è›‹ç™½çš„ç¼©å†™åç§°ï¼Œæ›¿æ¢åŸæ¥é¢„æµ‹çš„åŸºå› åç§°ï¼ˆæ¨¡ä»¿å®éªŒ 3 æ ¼å¼è½¬æ¢åçš„ GFF æ–‡æ¡£ï¼‰ï¼Œä¿å­˜ä¸ºä¸€ä¸ªæ–°çš„ GFF æ–‡æ¡£ã€‚

```bash
perl blast92gff3.pl Sc_blastx_gene_results.outfmt6 > Sc_blastx_gene_results.gff
# Summary of HSPs saved
# ALL saved = 4498
# other saved = 4498
```

##### ä¿¡æ¯æ•´åˆè„šæœ¬

```py
def getMapping(filePath):
    dfrm_gff = pd.read_csv(filePath, sep="\t", header=None, skiprows=1)
    unp_pattern = re.compile("sp:([A-z0-9_\|]+)")
    gene_pattern = re.compile("Target=([A-z0-9]+)")

    di = {}

    for index in dfrm_gff.index:
        gene = gene_pattern.search(dfrm_gff.loc[index, 8]).group(1)
        di[gene] = unp_pattern.search(dfrm_gff.loc[index, 0]).group(1)

    return di

def updateGFF(inPath, outPath, di):
    with open(inPath, "rt") as inFile:
        with open(outPath, "wt") as outFile:
            startwith = re.compile("# start gene ([A-z0-9]+)")
            flag = 0
            for line in inFile:
                startToken = startwith.search(line)
                if startToken is not None:
                    flag = 1
                    key = startToken.group(1)
                    outFile.write(line)
                    continue
                if flag:
                    line = line[:-1] + ";%s\n" % di.get(key, "")
                    flag = 0
                outFile.write(line)


di = getMapping("Sc_blastx_gene_results.gff")
addNota("Sc_augustus_out.gff3", "augustus_addNota.gff", di)
```

å˜é‡```di```å³å¯å…·ä½“æŸ¥çœ‹åŒ¹é…ä¸ŠåŸºå› çš„æƒ…å†µï¼Œè¿›è¡Œç»Ÿè®¡ã€‚

å¾—åˆ°åŠ å…¥äº†è›‹ç™½ç¼©å†™åç§°çš„augustusç»“æœ```augustus_addNota.gff```ã€‚

### 5. ä»å¤´é¢„æµ‹ç»“æœçš„è¯„ä¼°

#### 5.1 gffcompareå¯¹æ¯”

ä½¿ç”¨ gffcompare å·¥å…·æŠŠç¬¬ 4 æ­¥ç»“æœä¸ 1.1 æ­¥åŸå§‹ GFF æ•°æ®ä»¥åŠå®éªŒ 3 ç»“æœè¿›è¡Œæ¯”è¾ƒï¼Œ æŸ¥çœ‹ç»“æœï¼Œå¹¶åˆ†æå®ƒä»¬ä¹‹é—´çš„å¼‚åŒä¹‹å¤„ã€‚

```bash
gffcompare -V -r $gffFile ./augustus_addNota.gff -o ./Sc_augustus_out_addNota
```

å¾—åˆ°ä¸‹åˆ—æ–‡ä»¶ï¼š

```bash
Sc_augustus_out_addNota.annotated.gtf
Sc_augustus_out_addNota.augustus_addNota.gff.refmap
Sc_augustus_out_addNota.augustus_addNota.gff.tmap
Sc_augustus_out_addNota.loci
Sc_augustus_out_addNota.stats
Sc_augustus_out_addNota.tracking
```

#### 5.2 gffcompareç»“æœè§£æ

æå–```Sc_augustus_out_addNota.stats```å†…å®¹å¦‚ä¸‹ï¼š

|| Sensitivity | Precision  |
-|-|-
Base level|    99.5     |    95.6    |
Exon level|    93.7     |    88.8    |
Intron level|    72.4     |    50.1    |
Intron chain level|    71.9     |    52.4    |
Transcript level|    95.4     |    92.0    |
Locus level|    95.5     |    92.0    |
-|-|-
Matching intron chains|164|
Matching transcripts|4956|
Matching loci|4955|
-|-|-
Missed exons|48/5434|(0.9%)
Novel exons|333/5730|(5.8%)
Missed exons|21/239|(8.8%)
Missed exons|125/345|(36.2%)
Missed exons|0/5187|(0.0%)
Missed exons|251/5385|(4.7%)


![fig](./figs/GFF.png "Fig of GFF")

å¯ä»¥çŸ¥é“ï¼ŒGFFæ–‡æ¡£ç›¸å½“äºåœ¨æ£€éªŒBLASTæ¯”å¯¹æ‰¾åˆ°çš„UniProtä¸åŸºå› ç»„GFFæ³¨é‡Šæ–‡æ¡£é‡Œçš„UniProtçš„ç»“æœæ¥è¿‘ç¨‹åº¦ã€‚ä¸”è§†åŸºå› ç»„GFFæ³¨é‡Šæ–‡æ¡£å†…å®¹çš†ä¸ºçœŸã€‚

* Base Levelï¼šåœ¨ç›¸åŒåæ ‡ä¸ŠæŠ¥å‘Šçš„å¤–æ˜¾å­ç¢±åŸºçš„æ•°ç›®æƒ…å†µ
  * Sensitivityï¼šé«˜è¾¾99.5%ï¼Œè¯´æ˜BLASTç»“æœåœ¨è¯¥æ°´å¹³ä¸Šç»“æœæ‰¾åˆ°äº†ç»å¤§éƒ¨åˆ†åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£ä¸­çš„å†…å®¹ï¼Œæå°‘æ•°ç¢±åŸºæ ¹æœ¬æ²¡æœ‰è¢«ä»»ä½•é¢„æµ‹çš„è½¬å½•æœ¬ï¼ˆtransfragsï¼‰å¤–æ˜¾å­æ‰€è¦†ç›–
  * Precisionï¼šè¾¾95.6%ï¼Œè¯´æ˜BLASTç»“æœåœ¨è¯¥æ°´å¹³ä¸Šæœ‰ä¸€å°éƒ¨åˆ†ï¼ˆ4.4%ï¼‰ç¢±åŸºè¢«é¢„æµ‹çš„è½¬å½•æœ¬å¤–æ˜¾å­è¦†ç›–ä½†æœªè¢«ä»»ä½•å‚è€ƒè½¬å½•æœ¬å¤–æ˜¾å­è¦†ç›–
* Exon levelï¼šä¸¤æ–‡ä»¶åŸºå› ç»„ä¸Šçš„å¤–æ˜¾å­é—´éš”äº¤é›†æƒ…å†µ
  * å¯ä»¥çœ‹åˆ°é¢„æµ‹åŸºå› ç»“æœçš„å¤–æ˜¾å­ä¸åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£çš„å¤–æ˜¾å­è¾¹ç•Œæœ‰ä¸€å®šå°å·®å¼‚
* Intron levelï¼šå†…å«å­é—´éš”
  * é¢„æµ‹åŸºå› çš„å†…å«å­è¾¹ç•Œæœ‰ä¸å°‘ä¸åŸºå› ç»„æ³¨é‡Šæ–‡æ¡£å­˜åœ¨å·®å¼‚ï¼Œä¸”é”™è¯¯é¢„æµ‹äº†æ›´å¤šå†…å«å­ï¼ŒPrecisionä»…50.1%
* Transcript levelï¼šé¢„æµ‹è½¬å½•æœ¬ä¸å‚è€ƒè½¬å½•æœ¬é—´çš„åŒ¹é…æƒ…å†µ
  * è½¬å½•æ°´å¹³æ˜¯åŒ¹é…è‰¯å¥½ï¼Œä½†ä¹Ÿæœ‰å°‘æ•°"è¯¯å·®"
* Locus levelï¼šè§‚å¯Ÿåˆ°çš„åŸºå› åº§ï¼ˆå¤–æ˜¾å­é‡å çš„è½¬å½•ç‰©ç°‡ï¼‰ä¸æ„å»ºçš„å‚è€ƒåŸºå› åº§çš„ç›¸ä¼¼åŒ¹é…æƒ…å†µ
  * åŸºå› åº§ä½ç½®ä¹ŸåŒ¹é…è‰¯å¥½ï¼Œä½†ä¹Ÿæœ‰å°‘æ•°"è¯¯å·®"

#### 5.3 ä¸å®éªŒä¸‰/åŒæºé¢„æµ‹ç»“æœè¿›è¡Œå¯¹æ¯”

å†æ¥çœ‹å®éªŒ3çš„Perlè„šæœ¬è¿è¡Œåçš„åŒæºé¢„æµ‹ç»“æœGFFæ–‡æ¡£
  || Sensitivity | Precision  |
  -|-|-
  Base level|	76.7|	96.3
  Exon level|	46.5|	30.5
  Intro level|	0.5|	0.1
  Intro chain level|	0.6|	0.1
  Transcript level|	60.2|	9.6
  Locus level|	60.3|	56.6
  -|-|-
  Matching intron chains|1|
  Matching transcripts|2179|
  Matching loci|2179|
  -|-|-
  Missed exons|	159/3803|(4.2%)
  Novel exons|	587/17753|(3.3%)
  Missed introns|	160/184|(87.0%)
  Novel introns|	1096/1151|(95.2%)
  Missed loci|	0/3611|(0.0%)
  Novel loci|	223/3852|(5.8%)

  å¯ä»¥æ˜æ˜¾çœ‹åˆ°åŒæºé¢„æµ‹çš„å„é¡¹æ°´å¹³(é™¤äº†BaseLevelçš„Precisionå¾ˆç†æƒ³)éƒ½ä¸å¦‚ä»å¤´é¢„æµ‹ã€‚ä»å¤´é¢„æµ‹çš„è¯¯å·®æ°´å¹³æ›´ä½ï¼Œæ¯”åŒæºé¢„æµ‹æ›´æœ‰ä¿¡æœåŠ›ã€‚

  å…·ä½“æ¥çœ‹ï¼Œå¯ä»¥å‘ç°ä»å¤´é¢„æµ‹èƒ½å¤Ÿå‘ç°æ›´å¤šçš„lociï¼Œæ›´å‡†ç¡®çš„å†…å«å­è¾¹ç•Œã€‚

  åŒæºé¢„æµ‹å› ä¸ºæ˜¯åˆ©ç”¨ç›¸ä¼¼æ€§åº¦é‡æ¥æŸ¥æ‰¾åºåˆ—ï¼Œå†…å«å­. å¤–æ˜¾å­è¾¹ç•Œæ— æ³•å‡†ç¡®å¤„ç†ï¼Œä¸”è®¡ç®—è¿‡ç¨‹ä¸­æŸ¥æ‰¾çš„æ˜¯fungiçš„è›‹ç™½æ•°æ®ï¼Œä¼šäº§ç”Ÿä¸å°‘è¿œç¼˜ç‰©ç§çš„ç»“æœï¼›åŒæ—¶ä¹Ÿæ²¡æ³•è§£å†³ç‰©ç§ç‰¹æœ‰çš„åŸºå› ä¸è›‹ç™½ï¼Œå› æ­¤æ€»ä½“ä¸Šçš„æ•ˆæœ(Sensitivity)ä¸å¦‚ä»å¤´é¢„æµ‹ã€‚

## å®éªŒé¡¹ç›®5: åŸºå› ç»„æ³¨é‡Šä¹‹å¯åŠ¨å­åˆ†æå’Œé¢„æµ‹

### å®éªŒç›®çš„

1. åŠ æ·±å¯¹åŸºå› å¯åŠ¨å­çš„ç†è§£å’Œè®¤çŸ¥
2. å­¦ä¼šå¦‚ä½•è·å–å·²çŸ¥åŸºå› çš„å¯åŠ¨å­åºåˆ—æ•°æ®
3. ç†Ÿæ‚‰`EPD`å’Œ`TransFac`æ•°æ®åº“çš„ä½¿ç”¨
4. å­¦ä¼šä½¿ç”¨å·²çŸ¥çš„å¯åŠ¨å­å’Œè½¬å½•å› å­TFBSsçš„HMMæ¨¡å‹ï¼Œå¹¶èƒ½å¤Ÿç‹¬ç«‹ç¼–ç¨‹åˆ©ç”¨è¯¥HMMæ¨¡å‹æ¥è®¡ç®—é‰´åˆ«æœªçŸ¥å¯åŠ¨å­

### å®éªŒæµç¨‹

```viz
digraph flowchart_5 {
    # rankdir=LR;
    fontname="Courier New";
    size="6,6"; ratio = fill;
    node [style="filled,setlinewidth(3)", color="#8383cc", fontname="Courier New", shape="Mrecord",fixedsize=true,width=2.5,fillcolor="#d9e7ee"];
    edge [color="0.635 0.707 0.707"];
    label="åŸºå› ç»„æ³¨é‡Šä¹‹å¯åŠ¨å­åˆ†æå’Œé¢„æµ‹";
    step1[label="æ•°æ®å‡†å¤‡"];
    step2[label="å¯åŠ¨å­å…ƒä»¶HMMæ•°æ®"];
    # step3[label="DNAå…ƒä»¶çš„è®¡ç®—é‰´åˆ«"];
    step4[label="ä¸åŸGFFæ–‡ä»¶è¿›è¡Œå¯¹æ¯”"];
    step5[label="åˆ†å€¼ç»“æœå¯è§†åŒ–"];
    step6[label="ROCæ›²çº¿ç»˜åˆ¶"];
    step7[label="æ®æœ€ä½³é˜ˆå€¼è¿›ä¸€æ­¥ç­›é€‰"];

    subgraph cluster_1{
        style=filled;
        color=lightgrey;
        node [color=white];
        label="DNAå…ƒä»¶çš„è®¡ç®—é‰´åˆ«"
        sub_c1_1[label="è®¡ç®—åŸå§‹å¾—åˆ†"];
        sub_c1_2[label="bootstrap & shuffle"];
        sub_c1_3[label="æ ¹æ®på€¼è¿›è¡Œè¿‡æ»¤"];
        sub_c1_1->sub_c1_2->sub_c1_3;
    }
    step1->step2->sub_c1_1;
    sub_c1_3->step4->step5->step6->step7->sub_c1_3;
}
```

### 1. æ•°æ®å‡†å¤‡

> å®éªŒé¡¹ç›® 1 ä¸­ä¸‹è½½çš„åŸºå› ç»„åºåˆ—ï¼ˆFASTA æ ¼å¼ï¼‰. ä»¥åŠç›¸åº”çš„ GFF æ ¼å¼æ³¨é‡Šæ–‡ä»¶

```bash
workDir="/home/student/s24/zeFengZhu/Gen/"
fastaFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.fna"
gffFile="$workDir/GCA_000977265.3_Sc_YJM1342_v1_genomic.gff"
```

### 2. å¯åŠ¨å­ç›¸å…³DNAå…ƒä»¶HMMæ•°æ®

#### 2.1 ä»`EPD`æ•°æ®åº“ä¸­ä¸‹è½½ä»»æ„ä¸€ç§å¯åŠ¨å­ç›¸å…³çš„DNAå…ƒä»¶çš„HMMæ•°æ®

* [Link](<https://epd.epfl.ch/promoter_elements.php> "Link")
* Promoter element HMMs derived from EPD release 68 (September 2001): TATA-box HMM trained from 900 unrelated general promoter sequences

```py
WEIGHT = {
    "A": [61, 16, 352, 3, 354, 268, 360, 222, 155, 56, 83, 82, 82, 68, 77],
    "C": [145, 46, 0, 10, 0, 0,	3, 2, 44, 135, 147, 127, 118, 107, 101],
    "G": [152, 18, 2, 2, 5, 0, 20, 44, 157,150, 128, 128, 128, 139, 140],
    "T": [31,309, 35, 374, 30, 121, 6, 121, 33, 48, 31, 52,	61,	75,	71]
    }
```

å¯ä»¥çœ‹åˆ°ï¼Œè¯¥æƒé‡çŸ©é˜µä¸­0å€¼è¾ƒå°‘ï¼Œé’ˆå¯¹æ€§æ¢ç©¶TATA-boxçš„MOTIFçš„èƒ½åŠ›ä¸æ˜¯ç‰¹åˆ«å¼ºã€‚

<table align="center">
  <tr>
    <td>
      <img src="./figs/Weight.png"></img>
    </td>
  </tr>
  <tr>
    <td>TATA-box HMM Matrix</td>
  </tr>
</table>

### 3 DNAå…ƒä»¶çš„è®¡ç®—é‰´åˆ«

#### 3.1 æ ¹æ®è¯¥ HMM æ•°æ®ï¼Œç¼–å†™ç¨‹åºå¯¹ä¸Šè¿°åŸºå› ç»„åºåˆ—è¿›è¡Œéå†ï¼Œè®¡ç®—åŸå§‹å¾—åˆ†

<table align="center">
  <tr>
    <td>
      <img src="./figs/HMMCode.png"></img>
    </td>
  </tr>
  <tr>
    <td>HMMCode</td>
  </tr>
</table>

```py
    # Part of my code
    @classmethod
    def seqIO(cls, path):
        name, seq = '', ''
        rows = cls.getFileRows(path)
        with open(path, 'rt') as seqFile:
            for index, line in enumerate(seqFile):
                if line.startswith(">"):
                    if index != 0:
                        yield name, seq
                        seq = ''
                    name = line[:-1]
                else:
                    seq += line[:-1].upper()
                    if index + 1 == rows:
                        yield name, seq

    def subSeq(self, str):
        for i in range(0, len(str)-self.length+1):
            yield i, str[i:i+self.length]
```

* ç¼–å†™å¦‚ä¸Šåºåˆ—éå†å‡½æ•°
* `seqIO()`ç”¨ä»¥äº§å‡ºç”Ÿæˆå™¨ï¼Œè¿­ä»£äº§å‡ºå•ç‹¬ä¸€æŸ“è‰²ä½“çš„å®Œæ•´åºåˆ—
* `subSet()`ä¹Ÿäº§å‡ºç”Ÿæˆå™¨ï¼Œè¿­ä»£äº§å‡ºæŒ‡å®šé•¿åº¦çš„æ»‘åŠ¨çª—å£å­åºåˆ—

```py
    # Part of my code
    def scoreSeq(self, seq, inb=False):
        score = 1
        for index, base in enumerate(seq):
            try:
                score *= self.weight[base][index]
            except KeyError:
                score = 0
                break
            except IndexError:
                print(seq, index, base)
                raise IndexError("Invalid length for inputed seq")

        if not inb:
            return score, self.bootstrapping(score, seq)
        else:
            return score
```

* ç¼–å†™æ‰“åˆ†å‡½æ•°å¦‚ä¸Š
* å¯¹è¾“å…¥çš„åºåˆ—éå†ï¼Œè®¿é—®æƒé‡çŸ©é˜µå“ˆå¸Œè¡¨(Python Dictionary)
* å¦‚è‹¥é‡åˆ°ATCGä»¥å¤–å­—ç¬¦ï¼Œåˆ™æ‰“åˆ†å³ä¸º0ã€‚

#### 3.2 åœ¨è®¡ç®—åˆ†å€¼çš„åŒæ—¶ï¼Œä½¿ç”¨ bootstrap æŠ½æ ·è¯„ä¼°çš„æ–¹æ³•å¯¹è®¡ç®—æ¯ä¸ªç‰‡æ®µå¯é æ€§på€¼

> æ¯éšæœºæ‰“ä¹±ï¼ˆshuffleï¼‰ä¸€æ¬¡å½“å‰ç‰‡æ®µï¼Œå°±è®¡ç®—ä¸€ä¸ªåˆ†å€¼ï¼Œå¹¶ä¸åŸå§‹å¾—åˆ†è¿›è¡Œæ¯”è¾ƒ

```py
from time import perf_counter
from collections import Counter
from scipy.special import comb
from numpy.random import shuffle
...
    # Part of my code
    def bootstrapping(self, score, seq):
        if score == 0:
            return None
        elif len(set(seq)) == 1:
            return 0
        else:
            pass

        recordDict = {}
        start = perf_counter()

        seq_len = len(seq)
        m = seq_len
        record, allPosNum = Counter(seq), 1
        for x in record.values():
            allPosNum *= comb(m, x)
            m -= x

        testNum = min(self.bootstrapNum, allPosNum)

        try:
            while len(recordDict) < testNum:
                seq_var = list(seq)
                self.seq = seq
                shuffle(seq_var)
                seq_var = ''.join(seq_var)
                if seq_var in recordDict.keys():
                    self.checkRunTime(start)
                    continue

                recordDict[seq_var] = self.scoreSeq(seq_var, inb=True)
                self.checkRunTime(start)
        except RuntimeError:
            print("skip")
            pass

        count = 0
        for varScore in record.values():
            if varScore > score:
                count += 1

        return count/testNum
```

* ç¼–å†™å¦‚ä¸Šbootstrappingå‡½æ•°
* ä¸è®¡ç®—åˆ†å€¼ä¸º0çš„åºåˆ—ï¼Œè¿”å›`None`
* æœ¬å‡½æ•°ä¼šé’ˆå¯¹è¾“å…¥åºåˆ—è®¡ç®—å…¶æ‰€æœ‰çš„æ‰“ä¹±åçš„æ’åºçš„æ•°ç›®(allPosNum)
  * è‹¥è¯¥æ•°ç›®å¤§äºæŒ‡å®šçš„bootstrappingæ¬¡æ•°(bootstrapNum)ï¼Œåˆ™è¿›è¡ŒæŒ‡å®šbootstrapNumæ¬¡æ•°çš„shuffleå»è®¡ç®—å¹¶å¾—å‡ºp-value
  * è‹¥è¯¥æ•°ç›®å°äºæŒ‡å®šçš„bootstrapNumï¼Œåˆ™æœ€å¤šè¿›è¡ŒallPosNumæ¬¡æ•°çš„shuffleï¼Œä½†æ­¤æ—¶çš„p-valueå·²ç»å¤±å»æ„ä¹‰ï¼Œå› æ­¤æŒ‡å®šçš„bootstrapNumä¸èƒ½è¿‡å¤§
* æ¯éšæœºæ‰“ä¹±(shuffle)ä¸€æ¬¡å½“å‰ç‰‡æ®µï¼Œå°±è®¡ç®—ä¸€ä¸ªåˆ†å€¼ï¼Œå¹¶ä¸åŸå§‹å¾—åˆ†è¿›è¡Œæ¯”è¾ƒ
* på€¼è®¡ç®—æ–¹æ³•: å‡è®¾æŠ½æ ·è¯„ä¼°æ€»æ¬¡æ•°ä¸ºNï¼Œæ¯æ¬¡è¯„ä¼°çš„ç‰‡æ®µå¾—åˆ†å¤§äºåŸå§‹å¾—åˆ†çš„æ¬¡æ•°ä¸ºn, åˆ™ p=n/N

#### 3.3 æ ¹æ® p å€¼å¤§å°è¿›è¡Œè¿‡æ»¤

> ç­›é€‰é˜ˆå€¼è‡³å°‘ä¸º 0.05ï¼Œä¿ç•™ p å€¼ä½äºé˜ˆå€¼çš„ç‰‡æ®µå¾—åˆ†. p å€¼. åŸºå› ç»„ä½ç½®. æ­£è´Ÿé“¾ç­‰ä¿¡æ¯

```py
import pandas as pd
...

    @staticmethod
    def filteringResult(pV, dict):
        return {chro: [(locus, score, pValue) for locus, score, pValue in value if pValue <= pV] for chro, value in dict.items()}

    def toDataFrame(self, dict, strand='+'):
        """convert the dict to a dataframe"""
        dfLyst = []
        allCols = ['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'pValue']
        for chro, data in dict.items():
            df = pd.DataFrame(data, columns=["start", "score", "pValue"])
            # df['attributes'] = df['pValue'].apply(lambda x: 'p-value={}'.format(x))
            df["seqid"] = ' '.join(chro[1:].split(' ')[0:1])
            df["strand"] = strand
            df["end"] = df["start"]+self.length-1
            for col in allCols:
                if col not in df.columns:
                    df[col] = '.'
            dfLyst.append(df)
        return pd.concat(dfLyst)[allCols]
```

* è¯¥å‡½æ•°æ ¹æ®æŒ‡å®šçš„p-valueè¿›è¡Œè¿‡æ»¤

### 4. æŠŠåˆ†æç»“æœä¸åŸºå› ç»„çš„æ³¨é‡Šä¿¡æ¯è¿›è¡Œå¯¹æ¯”

> åˆ†æè¿™äº›é¢„æµ‹ç»“æœä¸å·²çŸ¥åŸºå› çš„ä½ç½®å…³ç³»ï¼Œæ‰¾åˆ°æ¯ä¸€ä¸ªå…ƒä»¶ä¸‹æ¸¸æœ€é‚»è¿‘çš„åŸºå› è½¬å½•èµ·å§‹ä½ç‚¹ï¼Œæ³¨æ„æ­£è´Ÿé“¾çš„åŒº


#### 4.1 è¿è¡Œ3ä¸­ç¼–å†™çš„ä»£ç ï¼Œå¾—åˆ°åˆ†æç»“æœ

```bash
# -b: --bootstrapNum; -p: --pValue; -c: --chroNum; -r: --reverse
python PromoterHMM.py -f $fastaFile -b 50 -p 0.05 -c 16 -o ./ -r True
```

bootstrapNumå–50æ¬¡ï¼Œå·²ç»æµ‹è¯•è¿‡10æ¬¡ï¼Œ20æ¬¡ï¼Œ50æ¬¡çš„pValueåˆ†å¸ƒæ›´å¯é ã€‚(?)

å¾—åˆ°æ–‡ä»¶:

```bash
output_50.gff3 # æ­£é“¾
output_reverse_50.gff3 # è´Ÿé“¾
```

#### 4.2 ç¼–å†™ä»£ç è¿›è¡Œä½ç‚¹å¯è§†åŒ–

ä»£ç ä¸ºé™„ä»¶ä¸­çš„`AnalysisResult.py`ï¼ŒåŒæ—¶è´Ÿè´£å¯è§†åŒ–ã€‚

```bash
python AnalysisResult.py -i output_50.gff3,output_reverse_50.gff3 -r $gffFile -o ./ -s +-
```

å¯¹è¾“å‡ºçš„åˆ†å€¼è¿›è¡Œå¯è§†åŒ–ï¼Œæ¨ªè½´ä¸ºä½ç‚¹ï¼Œçºµè½´ä¸ºåˆ†å€¼ã€‚å›¿äºåŸºå› ä½ç‚¹ä»¥åŠåˆ†å€¼ä½ç‚¹è¾ƒå¤šï¼Œå³ä¾¿å±•ç¤ºä¸æœ€è¿‘åŸºå› çš„è·ç¦»ï¼Œä¹Ÿæ— æ³•æ­£ç¡®åˆ¤æ–­ï¼Œå› æ­¤ä¸‹å›¾æš‚ä¸è¡¨å¾ä¸åŸºå› çš„è·ç¦»ï¼Œä¸‹é¢çš„åˆ†æå°†ä¼šæåŠã€‚

##### æ­£é“¾

<table align="center">
  <tr>
    <td>
      <img src="./figs/siteScore_+.png"></img>
    </td>
  </tr>
  <tr>
    <td>siteScore_+</td>
  </tr>
</table>

##### è´Ÿé“¾

<table align="center">
  <tr>
    <td>
      <img src="./figs/siteScore_-.png"></img>
    </td>
  </tr>
  <tr>
    <td>siteScore_-</td>
  </tr>
</table>

##### é¢„æµ‹å…ƒä»¶æ•°ä¸åŸºå› æ•°çš„ç»Ÿè®¡

|æŸ“è‰²ä½“|æ­£è´Ÿé“¾|é¢„æµ‹å…ƒä»¶æ•°*|åŸºå› æ•°|
|---|---|---|---|
|all|+|12045|3048|
|all|-|12166|3849|
CP005447.2| + |884|242
CP004647.2| + |726|189
CP005038.2| + |358|97
CP005346.2| + |673|165
CP006294.2| + |354|70
CP004945.1| + |272|56
CP005643.1| + |1124|271
CP004465.2| + |180|44
CP005249.1| + |1072|283
CP005151.2| + |658|175
CP006398.1| + |1552|323
CP006082.2| + |601|132
CP004810.2| + |520|148
CP005549.2| + |675|205
CP006174.2| + |869|241
CP004715.2| + |152|367

* *(pValue<0.05)

#### 4.3 HMMåˆ†ç±»å™¨æ•ˆæœè¯„ä¼°

##### æ¥è‡ªYJM1342çš„gffæ–‡æ¡£

<table align="center">
  <tr>
    <td>
      <img src="./figs/gffFig.png"></img>
    </td>
  </tr>
  <tr>
    <td>YJM1342 GFF Fig</td>
  </tr>
</table>

ä¸‹å›¾ROCæ¥è‡ª`AnalysisResult.py`è¿è¡Œç»“æœã€‚å›¾ä¸ºè®¾å®šä¸€ä¸ªè·ç¦»é˜ˆå€¼100bpçš„+-é“¾è”åˆåˆ†ç±»ç»“æœï¼Œä¸å¯¹åº”é“¾ä»¥åŠå¯¹åº”æŸ“è‰²ä½“çš„exonçš„startä½ç‚¹ä¸Šæ¸¸è·ç¦»å¤„äºé˜ˆå€¼ä»¥å†…çš„è®¾å®šä¸ºé˜³æ€§ï¼Œä»¥å¤–çš„ä¸ºé˜´æ€§ï¼›ç„¶åä»¥è¯¥å…ƒä»¶çš„åˆ†å€¼æ¥ç»˜åˆ¶ROCæ›²çº¿ã€‚

<table align="center">
<tr>
    <td>
      <img src="./figs/label.png"></img>
    </td>
  </tr>
  <tr>
    <td>label Fig</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/ROC.png"></img>
    </td>
  </tr>
  <tr>
    <td>ROC Fig</td>
  </tr>
</table>

å¯ä»¥çœ‹åˆ°æ€»ä½“ROCå¹¶ä¸å‡ºè‰²ï¼ŒAUCä»…0.51ï¼Œè¯´æ˜æ­¤æ¬¡HMMåˆ†ç±»å™¨åœ¨è¯¥100bpé˜ˆå€¼æ¡ä»¶ä¸‹æ˜¯å¼±åˆ†ç±»å™¨ã€‚

##### è¿›ä¸€æ­¥æ¢ç©¶

åˆ†ç±»æ•ˆæœå¹¶ä¸ç†æƒ³ï¼ŒçŒœæµ‹å¯èƒ½æ˜¯100bpçš„é˜ˆå€¼æ¡ä»¶è®¾ç½®æœ‰æ‰€å½±å“ï¼Œå› æ­¤ä¸‹é¢æ¢ç©¶æœ€ä½³é˜ˆå€¼æ¡ä»¶è®¾ç½®ã€‚

å¯¹æ¯ä¸ªé¢„æµ‹å…ƒä»¶è®¡ç®—å…¶ä¸‹æ¸¸çš„æœ€è¿‘gene startä½ç‚¹(å¯¹åº”é“¾. å¯¹åº”æŸ“è‰²ä½“)ä¸ä¹‹çš„è·ç¦»ï¼Œå–æœ€è¿‘çš„è¿›è¡Œç»Ÿè®¡ï¼Œåˆ†å¸ƒå›¾å¦‚ä¸‹ï¼š

<table align="center">
  <tr>
    <td>
      <img src="./figs/min_dis.png"></img>
    </td>
  </tr>
  <tr>
    <td>min_dis Fig</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/siteScatter.png"></img>
    </td>
  </tr>
  <tr>
    <td>siteScatter Fig</td>
  </tr>
</table>

å¯ä»¥çœ‹åˆ°è·ç¦»åˆ†å¸ƒä¸»è¦é›†ä¸­åœ¨1000bpä»¥å†…ï¼ŒåŒæ—¶éƒ¨åˆ†æŸ“è‰²ä½“çš„é¢„æµ‹ç»“æœä¸å°‘ç‰¹åˆ«ç†æƒ³ï¼Œé¢„æµ‹å…ƒä»¶è·ç¦»åŸºå› å¤ªè¿œ,å†æŸ¥çœ‹1000bpå†…çš„åˆ†å¸ƒå¦‚ä¸‹:

<table align="center">
  <tr>
    <td>
      <img src="./figs/min_dis2.png"></img>
    </td>
  </tr>
  <tr>
    <td>min_dis detail Fig</td>
  </tr>
</table>

ç”±æ­¤å›¾æ¥çœ‹ï¼Œç›®å‰è¿˜æ— æ³•åˆ¤æ–­æœ€ä½³é˜ˆå€¼ï¼Œå› æ­¤å¯¹å¤šä¸ªé˜ˆå€¼è¿›è¡Œæµ‹è¯•ï¼Œç»˜åˆ¶ROCï¼Œè®¡ç®—AUCï¼Œåˆ¤æ–­åˆé€‚é˜ˆå€¼ï¼š

<table align="center">
  <tr>
    <td>
      <img src="./figs/ROC_multiNew.png"></img>
    </td>
  </tr>
  <tr>
    <td>ROC_multi Fig</td>
  </tr>
</table>

å°½ç®¡åœ¨1000bpå†…ä»¥5bpä¸ºæ­¥é•¿ï¼Œæµ‹è¯•äº†å¤šä¸ªé˜ˆå€¼ï¼Œå¾—å‡º200ä»¥å†…æ‰æœ‰å¤§äº0.52çš„é˜ˆå€¼ï¼Œä½†æ˜¯AUCä»æ— æ˜æ˜¾é«˜å€¼ã€‚

ç©¶å…¶åŸå› ä¾¿æ˜¯scoreæ‰“åˆ†æ— æ³•ç²¾ç¡®å°†é˜³æ€§ä¸é˜´æ€§æ•°æ®åŒºåˆ†å¼€æ¥ï¼Œç›¸å½“äºéšæœºé€‰æ‹©ã€‚è¿½æ ¹æº¯æºä¾¿æ˜¯æœ¬å®éªŒé‡‡ç”¨çš„HMMçŸ©é˜µæ— æ³•æ­£ç¡®åŒºåˆ†å‡ºå¯åŠ¨å­ï¼ŒYJM1342çš„å¯åŠ¨å­å…ƒä»¶ä¸HMMç»“æœçŸ©é˜µçš„TATAè®­ç»ƒé›†ä¸å»åˆï¼Œä¸”ä»çŸ©é˜µçš„å„ä¸ªç¢±åŸºåˆ†å€¼æ¥çœ‹ï¼Œå…¶ä¸­0å€¼ä»¥åŠè¾ƒä½å€¼çš„å‚æ•°è¾ƒå°‘ï¼Œåˆ†å€¼æ™®éè¾ƒé«˜ï¼Œè¿™å°±å¯¼è‡´äº†å¤§é‡å‡é˜³æ€§çš„é¢„æµ‹å…ƒä»¶ï¼›åŒæ—¶å‡é˜³æ€§å…ƒä»¶ä¸­è¿˜å­˜åœ¨è¾ƒé«˜åˆ†å€¼ã€‚ä¸‹å›¾ä¸ºé˜ˆå€¼å–10bpçš„é˜³æ€§é˜´æ€§åˆ†å€¼åˆ†å¸ƒå›¾ã€‚

<table align="center">
  <tr>
    <td>
      <img src="./figs/Class.png"></img>
    </td>
  </tr>
  <tr>
    <td>Classify Fig</td>
  </tr>
</table>

æœ€ç»ˆé‡‡å–é˜ˆå€¼ä¸º10bpï¼Œæ ¹æ®ROCå¾—åˆ°çš„æœ€ä½³scoreé˜ˆå€¼ä¸º0.009687(logå€¼ä¸º-4.637)

#### 4.4 æ ¹æ®ä¸Šä¸€æ­¥è®¡ç®—çš„é˜ˆå€¼ï¼Œå¯¹ç¬¬3æ­¥çš„ç»“æœè¿›è¡Œè¿›ä¸€æ­¥çš„ç­›é€‰ï¼Œå¹¶æŒ‰ç…§ GFF3 æ ¼å¼ä¿å­˜

```py
df[(df["pValue"]<=0.05) & (df['score']>=0.009687)].to_csv("output_50_TATA_selected.gff3", sep="\t", index=False, header=False)
df_reverse[(df_reverse["pValue"]<=0.05) & (df_reverse['score']>=0.009687)].to_csv("output_reverse_50_TATA_selected.gff3", sep="\t", index=False, header=False)
```

## å®éªŒé¡¹ç›®6: åŸºå› ç»„å¯è§†åŒ–

### å®éªŒç›®çš„

1. åŠ æ·±å…¨åŸºå› ç»„æ³¨é‡Šä¿¡æ¯å¯è§†åŒ–æ„ä¹‰çš„ç†è§£
2. ç†Ÿæ‚‰å’ŒæŒæ¡å¸¸ç”¨å¯è§†åŒ–å·¥å…·çš„ä½¿ç”¨
3. åŠ å¼ºè‡ªä¸»å­¦ä¹ èƒ½åŠ›
4. åŠ æ·±ç»Ÿè®¡å­¦åœ¨åŸºå› ç»„æ•°æ®åˆ†æä¸­çš„åº”ç”¨
5. åŸ¹å…»å‘ç°é—®é¢˜. åˆ†æé—®é¢˜å’Œè§£å†³é—®é¢˜çš„èƒ½åŠ›

### 1. åŸºå› ç»„æµ‹åºå¯è§†åŒ–å·¥å…·ç›¸å…³æ–‡çŒ®èµ„æ–™çš„è°ƒç ”

> é€šè¿‡æ£€ç´¢å…¬å…±æœç´¢å¼•æ“æˆ–ä¸“ä¸šæ•°æ®åº“ï¼ˆPubMedï¼‰ï¼ŒæŸ¥é˜…1ç§åŸºå› ç»„å¯è§†åŒ–å·¥å…·æˆ–å®ä¾‹ï¼Œå¹¶å¯¹å…¶åŠŸèƒ½ç‰¹å¾è¿›è¡Œæè¿°

#### 1.1 Search in `Pubmed`

keyword:

```keyword
Genom* Browser[tiab]
Genom* View* [tiab]
```

#### 1.2 `USCS Genome Browser`

> Karolchik D, Hinrichs AS, Kent WJ. The UCSC Genome Browser. Current protocols in bioinformatics / editoral board, Andreas D Baxevanis [et al]. 2009;Chapter 1:Unit1 4. Epub 2009/12/04. 10.1002/0471250953.bi0104s28 PubMed Central PMCID: PMC2834533.

1. ç”±åŠ åˆ©ç¦å°¼äºšå¤§å­¦åœ£å…‹é²æ–¯åˆ†æ ¡ï¼ˆUCSCï¼‰å¼€å‘çš„åŸºå› ç»„æµè§ˆå™¨(genome.ucsc.edu)
2. å¯ä»¥å¿«é€Ÿæ˜¾ç¤ºä»»æ„æ¯”ä¾‹çš„åŸºå› ç»„çš„æ‰€éœ€éƒ¨åˆ†ï¼Œå¹¶é™„å¸¦ä¸€ç³»åˆ—å¯¹é½çš„æ³¨é‡Šâ€œè½¨è¿¹â€ã€‚
3. ç”±UCSCåŸºå› ç»„ç”Ÿç‰©ä¿¡æ¯å­¦å°ç»„å’Œå¤–éƒ¨åˆä½œè€…ç”Ÿæˆçš„æ³¨é‡Šæ˜¾ç¤ºäº†åŸºå› é¢„æµ‹ï¼ŒmRNAå’Œè¡¨è¾¾çš„åºåˆ—æ ‡ç­¾æ¯”å¯¹ï¼Œç®€å•çš„æ ¸è‹·é…¸å¤šæ€æ€§ï¼Œè¡¨è¾¾å’Œè°ƒæ§æ•°æ®ï¼Œè¡¨å‹å’Œå˜å¼‚æ•°æ®ä»¥åŠæˆå¯¹å’Œå¤šç‰©ç§æ¯”è¾ƒåŸºå› ç»„å­¦æ•°æ®ã€‚
4. ä¸ä¸€ä¸ªåŒºåŸŸæœ‰å…³çš„æ‰€æœ‰ä¿¡æ¯éƒ½æ˜¾ç¤ºåœ¨ä¸€ä¸ªçª—å£ä¸­ï¼Œä»¥æ–¹ä¾¿è¿›è¡Œç”Ÿç‰©å­¦åˆ†æå’Œè§£é‡Šã€‚
5. å¯ä»¥ä½¿ç”¨å¦ä¸€ä¸ªåŸºäºWebçš„åº”ç”¨ç¨‹åºUCSCè¡¨æµè§ˆå™¨æŸ¥çœ‹ï¼Œä¸‹è½½å’Œæ“ä½œGenome Browserè½¨é“ä¸‹é¢çš„æ•°æ®åº“è¡¨ã€‚ç”¨æˆ·å¯ä»¥åœ¨ä¸¤ä¸ªæµè§ˆå™¨ä¸­å°†æ•°æ®ä½œä¸ºè‡ªå®šä¹‰æ³¨é‡Šè½¨é“ä¸Šè½½ï¼Œä»¥ä¾›ç ”ç©¶æˆ–æ•™è‚²ç”¨é€”ã€‚

### 2. åŸºå› ç»„å¯è§†åŒ–

> åˆ©ç”¨å…¶ä¸­ä»»æ„ä¸€ä¸ªå¯ä»¥æœ¬åœ°åŒ–çš„å·¥å…·å¯¹å®éªŒä¸€æ‰€é€‰ç‰©ç§åŸºå› ç»„å’Œæœæœ‰æ³¨é‡Šä¿¡æ¯è¿›è¡Œå¯è§†åŒ–ï¼›å¯è§†åŒ–æˆåŠŸåï¼Œä»»é€‰ä¸€ä¸ªåŒ…å«æ³¨é‡Šä¿¡æ¯çš„åŒºé—´æˆªå›¾ä¿å­˜

#### 2.1 ä¸‹è½½IGV

#### 2.2 è¿›è¡ŒåŸºå› ç»„æœ¬åœ°å¯è§†åŒ–

##### 2.2.1 å¯¼å…¥åŸºå› ç»„

```bash
# Genomes -> Lode Genome from file
GCA_000977265.3_Sc_YJM1342_v1_genomic.fna
```

##### 2.2.2 å¤„ç†åŸGFFæ³¨é‡Šæ–‡ä»¶

ç”±äº`IGV`å¯è§†åŒ–æ—¶ä¼šå°†regionåŒºåŸŸè¿æ¥ï¼Œä½¿å¾—å„ä¸ªåŸºå› è§†å›¾ä¸Šç»„åˆåœ¨ä¸€èµ·ï¼Œä¸åˆ©äºæ­£ç¡®åˆ¤æ–­åŸºå› æƒ…å†µï¼Œå› æ­¤éœ€è¦å¯¹åŸGFFæ–‡ä»¶è¿›è¡Œå¤„ç†ï¼Œåˆ é™¤æ‰regionç›¸å…³è¡Œã€‚

```py
import pandas as pd

path = "GCA_000977265.3_Sc_YJM1342_v1_genomic.gff"
out = "GCA_000977265.3_Sc_YJM1342_v1_genomic_modified.gff"
dfrm = pd.read_csv(path, sep="\t", skiprows=7, header=None)
dfrm[dfrm[2]!="region"].to_csv(out, sep="\t", index=False, header=False)
```

##### 2.2.3 å¯¼å…¥GFFæ³¨é‡Šæ–‡ä»¶

```bash
# File -> Lode from file
GCA_000977265.3_Sc_YJM1342_v1_genomic_modified.gff
# é‡‡ç”¨Perlä»£ç è½¬æ¢æ¥çš„åŒæºé¢„æµ‹(blast)ç»“æœçš„gff3æ–‡æ¡£
Sc_perl_modified.gff3
# é‡‡ç”¨Augustusä»å¤´é¢„æµ‹ç»“æœçš„gffæ–‡æ¡£
augustus_addNota.gff
```

<table align="center">
  <tr>
    <td>
      <img src="./figs/IGV.png"></img>
    </td>
  </tr>
  <tr>
    <td>æ€»è§ˆå›¾</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/IGV2.png"></img>
    </td>
  </tr>
  <tr>
    <td>å¤šå¤–æ˜¾å­ 1</td>
  </tr>
  <tr>
    <td>
      <img src="./figs/IGV3.png"></img>
    </td>
  </tr>
  <tr>
    <td>å¤šå¤–æ˜¾å­ 2</td>
  </tr>
</table>

å¯ä»¥çœ‹åˆ°ï¼Œåœ¨åŸºå› åº§ä½ä¸Šï¼Œä¸å®éªŒ4æ€»ç»“çš„ä¸€è‡´ï¼Œä»å¤´é¢„æµ‹ç»“æœæ¯”åŒæºé¢„æµ‹ç»“æœæ›´ä¸ºå‡†ç¡®ï¼Œèƒ½å¤Ÿæ›´å‡†ç¡®åœ°é¢„æµ‹åŸºå› ä½ç½®ï¼›åœ¨å¤šå¤–æ˜¾å­ç»“æ„ä¸Šï¼ŒåŒæºé¢„æµ‹ä¸ä»å¤´é¢„æµ‹éƒ½èƒ½æ­£ç¡®é¢„æµ‹å‡ºéƒ¨åˆ†å¤šå¤–æ˜¾å­ç»“æ„ï¼Œä½†æ˜¯åŒæºé¢„æµ‹å¯èƒ½ä¼šäº§ç”Ÿä¸€å®šé”™è¯¯é¢„æµ‹ç»“æ„,ä¸”ä»å¤´é¢„æµ‹ç»“æœæ›´å¥½ã€‚
