# Concepts Related to Epigenomics

> Last Modified Time 2019-09-05

## Epigenetic Landscape

* DNA-methylation
* Nucleosome position (MNase-Seq, ATAT-Seq)
* TF Binding (ChIP-Seq? DNase-Seq/FAIRE-Seq? ATAT-Seq?)
* Histone Modification
* Chromatin accessibility (ATAT-Seq)
* Higher order chromatin interactions
* Non-coding RNA
* ...

### DNA-methylation

* å¯Œé›†åŒºåŸŸ: repetitive elements (CpG island e.g)
* Transferase:
* Heritable

#### DNA-methylation Detection

> ...

#### CpG island

> GC-rich regions that...

* The promoter regions of ~70% of genes have CpG islands.

##### Promoter Region

> The 1kb(or 2kb) upstream and downstream of TSS(Transcription Start Site).

### Histone Modification

* ç”²åŸºåŒ–
* ç£·é…¸åŒ–
* ä¹™é…°åŒ–
* æ³›ç´ åŒ–
* ç£ºé…°åŒ–

#### Histone Code

Code | Description | Location
-|-|-
H3K9me3 | __Repress__ | Gene Body
H3K27ac | Activate | Enhancer
H3K27me3 | __Repress__ |Promoter(@ EZH2 é…¶+ DNAç”²åŸºåŒ–é…¶)
H3K7?me3 | __Repress__ |?
H3K4me3 | Activate | Promoter
H3K9ac | Activate | ?
H3K4me2 | Activate | Promoter,Enhancer

* Gene Body: H3K9me3(Repress)
* Promoter: H3K27me3(Repress) H3K4me2/3(Activate)
* Enhancer: H3K27ac(Activate) H3K4me2/1(Acivate)

> ä½äºç»„è›‹ç™½Næœ«ç«¯, ä¸ä»…å­˜åœ¨äºH3ä¸Š

##### Combination of Histone Code

* H3K4me3 + H3K36me3(transcription elongation) $\rightarrow$ Activate

### Non-coding RNA

* microRNA
* é•¿é“¾éç¼–ç RNA

---

### ChIP-seq

> æŸ“è‰²è´¨å…ç–«å…±æ²‰æ·€

#### å®éªŒæ“ä½œ

1. Bacteria cell culture
2. Cross-linking (Fix with formaldehyde)
3. Sonication (DNA fragmentation)
4. Input (Only sonicated DNA)
5. Purification and quantification of ChIP-DNA

> å¯¹ç…§å¾ˆé‡è¦

å¯¹ç…§çš„ä½œç”¨ï¼š

* é™ä½èƒŒæ™¯å™ªå£°bias

#### æ•°æ®å¤„ç†

```reads``` -> ```alignment``` -> ```signal construction``` -> ```peak calling``` -> æ‰«motif -> ?

#### ç®—æ³•

* call peak: MACSç®—æ³• (æ³Šæ¾åˆ†å¸ƒç›¸å…³)
  * [link](https://www.jianshu.com/p/390f6d57488d "Link")
* reads å›å¸–
  * bowtie
  * tophat
  * BWA
* Tools for Motif Scan
  * MDScan
  * MEME
  * HOMER
  * MDSeqPos

> è®¡ç®—è½¬å½•æ°´å¹³: cufflinks

### TF

* Basal Transcription Regulation
* Differential enhancement of transcription
* Development
* Response to intercellular signals
* ...

#### Regulation

* Synthesis
* Nuclear localization
* Activation
* Accessibility of DNA-binding site
* Availability of other cofactors/transcription factors
* ...

#### TF Binding Motif DB / Motif Knowledgebase

* Jaspar

---
> 2019-10-10

1. ?
2. ?
3. ?
4. Conserve <-> functional ? => functional -> conserve
5. å¤šä¸ªTFç›¸äº’ä½œç”¨
6. Cell Type-Specific Binding

Concepts:

* på€¼
* æ§åˆ¶é”™è¯¯å‘ç”Ÿç‡ï¼šæ§åˆ¶på€¼æ›´ä½
* Qå€¼,æ ¡æ­£På€¼
* BHæ³•æ ¡æ­£På€¼

Info:

1. åŸºå› ç»„ç‰ˆæœ¬
2. ç‰©ç§
3. åŸºå› 
4. fastaqæ–‡ä»¶

### DNase-Seq, MNase-Seq

* DNase-Seq: identify the location of regulatory regions
  * é™åˆ¶æ€§å†…åˆ‡é…¶ (DNase I)
  * å®éªŒéš¾åš
  * é…¶åˆ‡æ—¶é—´ä¸å¥½æ§åˆ¶
  * ç»†èƒé‡è¦æ±‚å¤§ï¼Œä¹Ÿå¯å•ç»†èƒ
  * åœ¨æŸ“è‰²è´¨è‡´å¯†åŒºåŸŸï¼ŒDNAé“¾è¢«è‡´å¯†ç»“æ„å¾ˆå¥½åœ°ä¿æŠ¤èµ·æ¥ï¼Œä½¿å¾—å†…åˆ‡é…¶æ— æ³•æ¥è¿‘è¿™äº›åŒºåŸŸï¼Œåªèƒ½åˆ‡å‰²å¼€æ”¾åŒºåŸŸçš„DNAã€‚åŒæ ·ï¼Œåœ¨å¼€æ”¾åŒºåŸŸï¼Œç¼ ç»•åœ¨æ ¸å°ä½“ä¸Šçš„DNAè¢«æ ¸å°ä½“ç»“æ„æ‰€ä¿æŠ¤ï¼Œåªæœ‰æ ¸å°ä½“ä¹‹é—´çš„DNAåºåˆ—èƒ½å¤Ÿè¢«DNase Iåˆ‡å‰²ï¼Œè¿™äº›åŒºåŸŸå†…èƒ½å¤Ÿè¢«DNaseåˆ‡å‰²çš„ä½ç‚¹ä¹Ÿè¢«ç§°ä¸ºDHSï¼Œå³DNaseè¶…æ•æ„Ÿä½ç‚¹ï¼ˆDNase hypersensitive sitesï¼‰
* MNase-Seq: æ ¸å°ä½“æ’å¸ƒ,
  > __DNase-seq__ is DNase I digestion combined with high-throughput DNA sequencing to get sites of digestion at single base-pair resolution. Itâ€™s kind of the "yin" to MNase-seqâ€™s "yang", providing inferred info on chromatin-binding proteins that lie between hypersensitive sites, whereas __MNase-seq__ maps protected regions. Both methods require another technique, such as ChIP-seq, for identifications of bound elements, though. (<https://epigenie.com/three-ways-to-get-intimate-with-epigenetic-marks/)>
  * é™åˆ¶æ€§å¤–åˆ‡é…¶
  * å’ŒDNase-seqåŸç†ç±»ä¼¼ï¼Œæ¢æµ‹åŒºåŸŸäº’è¡¥ï¼›MNase-seqä½¿ç”¨çš„é…¶æ˜¯é™åˆ¶æ€§å¤–åˆ‡é…¶ï¼Œå°†ä¸å—ä¿æŠ¤çš„åŒºåŸŸç»Ÿç»Ÿåˆ‡é™¤ï¼Œåªä½™ä¸‹æ ¸å°ä½“ä¸Šç¼ ç»•çš„DNAåºåˆ—
  * MNase-Seqæ•°æ® ä¸èƒ½ç”¨Macs call peak
  * ç»†èƒé‡è¦æ±‚å¤§
* FAIRE-Seq
  > FAIRE-Seq (Formaldehyde-Assisted Isolation of Regulatory Elements) is a method in molecular biology used for determining the sequences of DNA regions in the genome associated with regulatory activity. The technique was developed in the laboratory of Jason D. Lieb at the University of North Carolina, Chapel Hill. In contrast to DNase-Seq, the FAIRE-Seq protocol doesn't require the permeabilization of cells or isolation of nuclei, and can analyse any cell type. In a study of seven diverse human cell types, __DNase-seq and FAIRE-seq produced strong cross-validation__, with each cell type having 1-2% of the human genome as open chromatin.
  * ä½¿ç”¨ç”²é†›å›ºå®š
  * ä¸chip-seqä¸€æ ·ä¹Ÿç”¨è¶…å£°æ¥æ‰“æ–­
  * ä¿¡å™ªæ¯”ä½

### ATAT-Seq

> å…¨ç§°Assay for Transposase Accessible Chromatin with high-throughput sequencingï¼Œå³åˆ©ç”¨è½¬åº§é…¶ç ”ç©¶æŸ“è‰²è´¨å¯è¿›å…¥æ€§çš„é«˜é€šé‡æµ‹åºæŠ€æœ¯ã€‚

* ATAC-seqå‡ºæ¥çš„ç»“æœï¼Œå’Œä¼ ç»Ÿæ–¹æ³•å‡ºæ¥çš„ç»“æœå…·æœ‰å¾ˆå¼ºçš„ä¸€è‡´æ€§
* åŒæ—¶ä¹Ÿå’ŒåŸºäºç»„è›‹ç™½ä¿®é¥°markerçš„ChIP-seqæœ‰è¾ƒé«˜çš„å»åˆç¨‹åº¦
* è€Œç›¸æ¯”èµ·æ¥ï¼ŒATAC-seqçš„é‡å¤æ€§ï¼Œæ¯”MNase-seqå’ŒDNase-seqçš„æ›´å¼ºï¼Œæ“ä½œèµ·æ¥ä¹Ÿæ›´åŠ ç®€å•ï¼Œè€Œä¸”åªéœ€è¦å¾ˆå°‘çš„ç»†èƒ/ç»„ç»‡é‡
* åŒæ—¶å‡ºæ¥çš„ä¿¡å·æ›´åŠ æ¼‚äº®
* ç›®å‰å·²ç»æ˜¯ç ”ç©¶æŸ“è‰²è´¨å¼€æ”¾æ€§é¦–é€‰çš„æŠ€æœ¯æ–¹æ³•ã€‚
* (<https://www.zhihu.com/question/263776928/answer/273229159)>

* åŒæ—¶å…·æœ‰MNase-Seqå¾—åˆ°çš„æ ¸å°ä½“æ’å¸ƒä¿¡æ¯ä»¥åŠå¯ç”±DNase-Seq/FAIRE-Seqå¾—åˆ°è°ƒæ§åŒºåŸŸä¿¡æ¯
* è¦è€ƒè™‘å»é™¤çº¿ç²’ä½“çš„é—®é¢˜

ATAC-Seqä¸ChIP-Seqçš„ä¸åŒçš„æ˜¯ATAC-Seqæ˜¯å…¨åŸºå› ç»„èŒƒå›´å†…æ£€æµ‹æŸ“è‰²è´¨çš„å¼€æ”¾ç¨‹åº¦ï¼Œå¯ä»¥å¾—åˆ°å…¨åŸºå› ç»„èŒƒå›´å†…çš„è›‹ç™½è´¨å¯èƒ½ç»“åˆçš„ä½ç‚¹ä¿¡æ¯ï¼Œä¸€èˆ¬ç”¨äºä¸çŸ¥é“ç‰¹å®šçš„è½¬å½•å› å­ï¼Œç”¨æ­¤æ–¹æ³•ä¸å…¶ä»–æ–¹æ³•ç»“åˆç­›æŸ¥æ„Ÿå…´è¶£çš„ç‰¹å®šè°ƒæ§å› å­ï¼›ä½†æ˜¯ChIP-Seqæ˜¯æ˜ç¡®çŸ¥é“æ„Ÿå…´è¶£çš„è½¬å½•å› å­æ˜¯ä»€ä¹ˆï¼Œæ ¹æ®æ„Ÿå…´è¶£çš„è½¬å½•å› å­è®¾è®¡æŠ—ä½“å»åšChIPå®éªŒæ‹‰DNAï¼ŒéªŒè¯æ„Ÿå…´è¶£çš„è½¬å½•å› å­æ˜¯å¦ä¸DNAå­˜åœ¨ç›¸äº’ä½œç”¨ã€‚ATAC-Seqã€ChIP-Seqã€Dnase-Seqã€MNase-Seqã€FAIRE-Seqæ•´ä½“çš„åˆ†ææ€è·¯ä¸€è‡´ï¼Œæ‰¾åˆ°å¯Œé›†åŒºåŸŸï¼Œå¯¹å¯Œé›†åŒºåŸŸè¿›è¡ŒåŠŸèƒ½åˆ†æã€‚

### Super-Enhancer

> 2019/11/28

#### Feature

* é«˜è¡¨è¾¾
* é«˜è°ƒæ§
* æ‹›å‹Ÿçš„TFæ•°é‡ä¼—å¤š
* å¤šä¸ªenhancerç›¸äº’åä½œ, ä¸”è¾ƒä¸ºå¯†é›†

#### Conclusion

> Embryonic stem cellï¼ŒESCs

* Large Genomic Domains Occupied by Master Transcription Factors and Mediator in ESCs
* Super-Enhancers Are Associated with Key ESC Identity Genes
* Super-Enhancers Are Cell-Type Specific and Mark Key Cell Identity Genes

## ReferenceğŸ‘©â€ğŸ’»

1. ğŸ“º [StatQuest: A gentle introduction to ChIP-Seq](https://www.youtube.com/watch?v=nkWGmaYRues)
2. ğŸ“º [MIT 7.91J: 7. ChIP-seq Analysis; DNA-protein Interactions](https://www.youtube.com/watch?v=Ob9xGBPvr_s)